<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Sean Davis">
<meta name="dcterms.date" content="2024-06-01">
<meta name="description" content="A collection of chapters more than a book, but….">
<title>24&nbsp; Microbiome analysis – The RBioc Book</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./bioc-summarizedexperiment.html" rel="next">
<link href="./geoquery.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-180dc90b24616f542b21b8b54fc3f545.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-5a15bea12966ac2249ff2e96cb8142e0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-KLLV1GCF4E"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-KLLV1GCF4E', { 'anonymize_ip': true});
</script><style>html{ scroll-behavior: smooth; }</style>
<script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">The RBioc Book</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./intro.html"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./data_structures_overview.html"> 
<span class="menu-text">Data Structures</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/seandavi/RBiocBook" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./geoquery.html">Bioconductor</a></li><li class="breadcrumb-item"><a href="./310_microbiome.html"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Microbiome analysis</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_to_rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r_intro_mechanics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R mechanics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Up and Running with R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages_and_dice.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Packages and more dice</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./reading_and_writing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Reading and writing data files</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./data_structures_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R Data Structures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./vectors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Vectors</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./matrices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lists.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lists</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dataframes_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Data Frames</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./factors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Factors</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./eda_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exploratory data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dplyr_intro_msleep.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to dplyr: mammal sleep dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eda_and_univariate_brfss.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Behavioral Risk Factor Surveillance System</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggplot2/intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Exploring data with <code>ggplot2</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualization_guide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Self-Guided Data Visualization in R</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./norm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Working with distribution functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./t-stats-and-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">The t-statistic and t-distribution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./kmeans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">K-means clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Machine Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine_learning/intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine_learning/models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Supervised Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine_learning/mlr3verse_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">The mlr3verse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine_learning/machine_learning_mlr3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Examples</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Bioconductor</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./geoquery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Accessing and working with public omics data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./310_microbiome.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Microbiome analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bioc-summarizedexperiment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Introduction to <code>SummarizedExperiment</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genomic_ranges_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Genomic Ranges Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ranges_and_signals.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Genomic ranges and features</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ranges_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Ranges Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./atac-seq/atac-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Genomic ranges and ATAC-Seq</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./single_cell/setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Single Cell RNA-seq Analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Appendix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git_and_github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Git and GitHub</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./additional_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Additional resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ai_tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">AI Tools for Enhanced Learning and Productivity in Bioinformatics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dataviz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">E</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./matrix_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">F</span>&nbsp; <span class="chapter-title">Matrix Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#getting-started" id="toc-getting-started" class="nav-link active" data-scroll-target="#getting-started"><span class="header-section-number">24.1</span> Getting started</a></li>
  <li><a href="#accessing-microbiome-datasets" id="toc-accessing-microbiome-datasets" class="nav-link" data-scroll-target="#accessing-microbiome-datasets"><span class="header-section-number">24.2</span> Accessing microbiome datasets</a></li>
  <li><a href="#data-containers" id="toc-data-containers" class="nav-link" data-scroll-target="#data-containers"><span class="header-section-number">24.3</span> Data containers</a></li>
  <li>
<a href="#loading-a-microbiome-dataset" id="toc-loading-a-microbiome-dataset" class="nav-link" data-scroll-target="#loading-a-microbiome-dataset"><span class="header-section-number">24.4</span> Loading a microbiome dataset</a>
  <ul class="collapse">
<li><a href="#experimental-measurements-assays" id="toc-experimental-measurements-assays" class="nav-link" data-scroll-target="#experimental-measurements-assays"><span class="header-section-number">24.4.1</span> Experimental measurements: <code>assays</code></a></li>
  <li><a href="#sample-information-coldata" id="toc-sample-information-coldata" class="nav-link" data-scroll-target="#sample-information-coldata"><span class="header-section-number">24.4.2</span> Sample information: <code>colData</code></a></li>
  <li><a href="#measured-feature-information-rowdata" id="toc-measured-feature-information-rowdata" class="nav-link" data-scroll-target="#measured-feature-information-rowdata"><span class="header-section-number">24.4.3</span> Measured feature information: <code>rowData</code></a></li>
  <li><a href="#rowtree" id="toc-rowtree" class="nav-link" data-scroll-target="#rowtree"><span class="header-section-number">24.4.4</span> rowTree</a></li>
  </ul>
</li>
  <li>
<a href="#wrangling-and-subsetting" id="toc-wrangling-and-subsetting" class="nav-link" data-scroll-target="#wrangling-and-subsetting"><span class="header-section-number">24.5</span> Wrangling and subsetting</a>
  <ul class="collapse">
<li><a href="#subsetting-samples" id="toc-subsetting-samples" class="nav-link" data-scroll-target="#subsetting-samples"><span class="header-section-number">24.5.1</span> Subsetting samples</a></li>
  <li><a href="#agglomerating-data" id="toc-agglomerating-data" class="nav-link" data-scroll-target="#agglomerating-data"><span class="header-section-number">24.5.2</span> Agglomerating data</a></li>
  </ul>
</li>
  <li>
<a href="#community-indices" id="toc-community-indices" class="nav-link" data-scroll-target="#community-indices"><span class="header-section-number">24.6</span> Community indices</a>
  <ul class="collapse">
<li><a href="#alpha-diversity" id="toc-alpha-diversity" class="nav-link" data-scroll-target="#alpha-diversity"><span class="header-section-number">24.6.1</span> Alpha diversity</a></li>
  <li><a href="#beta-diversity" id="toc-beta-diversity" class="nav-link" data-scroll-target="#beta-diversity"><span class="header-section-number">24.6.2</span> Beta diversity</a></li>
  </ul>
</li>
  <li>
<a href="#microbial-composition" id="toc-microbial-composition" class="nav-link" data-scroll-target="#microbial-composition"><span class="header-section-number">24.7</span> Microbial composition</a>
  <ul class="collapse">
<li><a href="#abundance" id="toc-abundance" class="nav-link" data-scroll-target="#abundance"><span class="header-section-number">24.7.1</span> Abundance</a></li>
  <li><a href="#prevalence" id="toc-prevalence" class="nav-link" data-scroll-target="#prevalence"><span class="header-section-number">24.7.2</span> Prevalence</a></li>
  </ul>
</li>
  <li><a href="#heatmaps" id="toc-heatmaps" class="nav-link" data-scroll-target="#heatmaps"><span class="header-section-number">24.8</span> Heatmaps</a></li>
  <li><a href="#further-fun" id="toc-further-fun" class="nav-link" data-scroll-target="#further-fun"><span class="header-section-number">24.9</span> Further fun</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/seandavi/RBiocBook/edit/main/310_microbiome.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/seandavi/RBiocBook/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/seandavi/RBiocBook/blob/main/310_microbiome.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./geoquery.html">Bioconductor</a></li><li class="breadcrumb-item"><a href="./310_microbiome.html"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Microbiome analysis</span></a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Microbiome analysis</span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source" data-quarto-source-url="https://github.com/seandavi/RBiocBook/blob/main/310_microbiome.qmd"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://seandavi.github.io/">Sean Davis</a> <a href="mailto:seandavi@gmail.com" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://medschool.cuanschutz.edu/">
            University of Colorado<br>Anschutz School of Medicine
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">June 1, 2024</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">July 7, 2025</p>
    </div>
  </div>
    
  </div>
  


</header><p>The human microbiome is a complex ecosystem of bacteria, viruses, fungi, and other microorganisms that live in and on the human body. These microorganisms play a crucial role in human health and disease, influencing everything from digestion and metabolism to immune function and mental health. Understanding the composition and function of the human microbiome is a rapidly growing area of research with implications for a wide range of fields, including medicine, agriculture, and environmental science.</p>
<p>Microbiome analysis involves studying the microbial communities that inhabit different environments, such as the human gut, skin, mouth, and soil. The analysis of microbiome data typically involves quantifying the abundance of different microorganisms in a given sample and comparing the microbial communities across different samples or conditions. This can help researchers identify patterns, relationships, and associations between microbial communities and various factors, such as health status, diet, lifestyle, and environmental conditions.</p>
<p>In this chapter, we will assume that the data has already been processed and cleaned, and we will focus on analyzing the microbiome data using R.</p>
<section id="getting-started" class="level2" data-number="24.1"><h2 data-number="24.1" class="anchored" data-anchor-id="getting-started">
<span class="header-section-number">24.1</span> Getting started</h2>
<p>We’ll be using the <code>mia</code> package for microbiome analysis in R. The <code>mia</code> package provides a suite of tools for analyzing microbiome data, including functions for loading, processing, and visualizing microbiome data. The package is designed to work with the <code>SummarizedExperiment</code> and <code>TreeSummarizedExperiment</code> classes, which are specialized data structures for storing microbiome data in R.</p>
<p>To get started, you’ll need to install the <code>mia</code> package from GitHub using the <code>BiocManager</code> package. If you don’t already have the <code>BiocManager</code> package installed, you can install it using the following command:</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Install the mia package</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"microbiome/mia"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Note:</strong> The <code>mia</code> package is part of the Bioconductor project, but in this case, we are installing it directly from GitHub using the <code>microbiome/mia</code> repository. This is a common practice when working with development versions of packages or with packages that are not yet available on Bioconductor.</p>
</div>
</div>
<p>Once the <code>mia</code> package is installed, you can load it into your R session using the following command:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load the mia package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that the <code>mia</code> package is loaded, we can load the microbiome dataset that we will be working with.</p>
</section><section id="accessing-microbiome-datasets" class="level2" data-number="24.2"><h2 data-number="24.2" class="anchored" data-anchor-id="accessing-microbiome-datasets">
<span class="header-section-number">24.2</span> Accessing microbiome datasets</h2>
<p>There are many publicly available microbiome datasets that you can use for analysis. Bioconductor provides several packages that contain curated microbiome datasets that facilitate access to and loading microbiome data into R for analysis.</p>
<p><a href="https://bioconductor.org/packages/curatedMetagenomicsData">curatedMetagenomicData</a> is a large collection of curated human microbiome datasets, provided as TreeSE objects (Pasolli et al.&nbsp;2017). The resource provides curated human microbiome data including gene families, marker abundance, marker presence, pathway abundance, pathway coverage, and relative abundance for samples from different body sites. See the package homepage for more details on data availability and access.</p>
<p><a href="https://bioconductor.org/packages/microbiomeDataSets">The microbiomeDataSets package</a> provides a collection of curated microbiome datasets for teaching and research purposes. The package contains several example datasets that can be used to explore different aspects of microbiome analysis, such as alpha and beta diversity, differential abundance analysis, and visualization.</p>
<p><a href="https://bioconductor.org/packages/mia">The mia package</a> and several other packages provide example datasets for learning and testing the functions in the package. These datasets are typically small and easy to work with, making them ideal for learning how to analyze microbiome data in R.</p>
</section><section id="data-containers" class="level2" data-number="24.3"><h2 data-number="24.3" class="anchored" data-anchor-id="data-containers">
<span class="header-section-number">24.3</span> Data containers</h2>
<p>Bioconductor provides several specialized data structures for storing and working with microbiome data in R. These data structures are designed to handle the complex and high-dimensional nature of microbiome data and provide a convenient and efficient way to store, manipulate, and analyze microbiome data in R.</p>
<p><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment (SE)</a> (Morgan et al.&nbsp;2020) is a generic and highly optimized container for complex data structures. It has become a common choice for analyzing various types of biomedical profiling data, such as RNAseq, ChIp-Seq, microarrays, flow cytometry, proteomics, and single-cell sequencing.</p>
<p><a href="https://bioconductor.org/packages/TreeSummarizedExperiment">TreeSummarizedExperiment</a> (Huang 2020) was developed as an extension to incorporate hierarchical information (such as phylogenetic trees and sample hierarchies) and reference sequences.</p>
</section><section id="loading-a-microbiome-dataset" class="level2 page-columns page-full" data-number="24.4"><h2 data-number="24.4" class="anchored" data-anchor-id="loading-a-microbiome-dataset">
<span class="header-section-number">24.4</span> Loading a microbiome dataset</h2>
<p>We will be using the <code>curatedMetagenomicData</code> package to load the <code>FengQ_2015</code> dataset. This dataset contains microbiome data from a study by <a href="https://pubmed.ncbi.nlm.nih.gov/25758642/">Feng et al.&nbsp;(2015)</a>. The curatedMetagenomicData package provides a convenient interface for accessing microbiome datasets that have been pre-processed and curated for analysis.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Take a moment to familiarize yourself with at least the title and abstract from the paper. We won’t be reproducing the analysis in the paper, but it is helpful to connect the code in this workflow to the biology and research.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Again, before using a package, we must install it</span></span>
<span><span class="co"># just once.</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">'curatedMetagenomicData'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once installed, we can use the package.</p>
<p>Ignore the details of the next few lines of code for now, but suffice it to say that what the code will do for us is to load a dataset for downstream analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load the FengQ_2015 dataset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/curatedMetagenomicData">curatedMetagenomicData</a></span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/curatedMetagenomicData.html">curatedMetagenomicData</a></span><span class="op">(</span><span class="st">"FengQ_2015.relative_abundance"</span>, dryrun <span class="op">=</span> <span class="cn">FALSE</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
$`2021-03-31.FengQ_2015.relative_abundance`
dropping rows without rowTree matches:
  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Atopobiaceae|g__Olsenella|s__Olsenella_profusa
  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Collinsella|s__Collinsella_stercoris
  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Enorma|s__[Collinsella]_massiliensis
  k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillales_unclassified|g__Gemella|s__Gemella_bergeri
  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Carnobacteriaceae|g__Granulicatella|s__Granulicatella_elegans
  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Ruminococcus|s__Ruminococcus_champanellensis
  k__Bacteria|p__Firmicutes|c__Erysipelotrichia|o__Erysipelotrichales|f__Erysipelotrichaceae|g__Bulleidia|s__Bulleidia_extructa
  k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales|f__Sutterellaceae|g__Sutterella|s__Sutterella_parvirubra
  k__Bacteria|p__Synergistetes|c__Synergistia|o__Synergistales|f__Synergistaceae|g__Cloacibacillus|s__Cloacibacillus_evryensis</code></pre>
</div>
</div>
<p>This code goes out to the internet and downloads the dataset for you. The <code>tse</code> object is a <code>TreeSummarizedExperiment</code> object that contains the microbiome data from the Feng et al.&nbsp;(2015) study. The <code>TreeSummarizedExperiment</code> class is a specialized data structure for storing microbiome data in R, and it is used by the <code>mia</code> package for microbiome analysis.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you are working with your own microbiome data, you will often need to load it into R using functions like <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> or <code><a href="https://rdrr.io/r/utils/read.table.html">read.table()</a></code> to read tabular data files. Make sure your data is properly formatted and cleaned before loading it into R for analysis.</p>
</div>
</div>
<p>Once the data is loaded into R, we can start exploring the data to understand its structure and contents. The <code>tse</code> object is a <code>TreeSummarizedExperiment</code> object, which is a specialized data structure for storing microbiome data in R. It works quite like a data.frame in, but with many specialized structures for storing microbiome experiment data as shown in <a href="#fig-tse-anatomy" class="quarto-xref">Figure&nbsp;<span>24.1</span></a>. We can use various functions to explore the data stored in the <code>tse</code> object. See the <a href="https://bioconductor.org/packages/TreeSummarizedExperiment">TreeSummarizedExperiment</a> package for more details.</p>
<p>This section provides an introduction to these data containers. In microbiome data science, these containers link taxonomic abundance tables with rich side information on the features (microbes) and samples. Taxonomic abundance data can be obtained by 16S rRNA amplicon or metagenomic sequencing, phylogenetic microarrays, or by other means. Many microbiome experiments include multiple versions and types of data generated independently or derived from each other through transformation or agglomeration.</p>
<div id="fig-tse-anatomy" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-tse-anatomy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/treesummarizedexperiment.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-tse-anatomy-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;24.1: Anatomy of a TreeSummarizedExperiment object. Compared to the SingleCellExperiment objects, TreeSummarizedExperiment has five additional slots. <strong>rowTree</strong>: the hierarchical structure on the rows of the assays. <strong>rowLinks</strong>: the link information between rows of the assays and the rowTree. <strong>colTree</strong>: the hierarchical structure on the columns of the assays. <strong>colLinks</strong>: the link information between columns of the assays and the colTree. <strong>referenceSeq</strong> (optional): the reference sequence data per feature (row).
</figcaption></figure>
</div>
<p>To get an overview of the data, we can simply type the name of the object in the R console:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Print the object</span></span>
<span><span class="va">tse</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 601 154 
metadata(1): agglomerated_by_rank
assays(1): relative_abundance
rownames(601): [Bacteroides] pectinophilus [Butyribacterium]
  methylotrophicum ... Weissella cibaria Weissella viridescens
rowData names(7): superkingdom phylum ... genus species
colnames(154): SID31004 SID31009 ... SID532832 SID532915
colData names(28): study_name subject_id ... ldl hba1c
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (601 rows)
rowTree: 1 phylo tree(s) (10430 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>Observe how the output of the <code>tse</code> object will show you the dimensions of the data, the metadata associated with the samples and features, etc. However, the data themselves are not printed to the console due to the large size of the dataset. The <code>TreeSummarizedExperiment</code> object is a complex data structure that contains multiple components, including the abundance data, sample information, feature information, and other metadata.</p>
<p>Next, we’ll look at some of the key components of the <code>TreeSummarizedExperiment</code> object and how to access and manipulate them.</p>
<section id="experimental-measurements-assays" class="level3" data-number="24.4.1"><h3 data-number="24.4.1" class="anchored" data-anchor-id="experimental-measurements-assays">
<span class="header-section-number">24.4.1</span> Experimental measurements: <code>assays</code>
</h3>
<p>When working with high-throughput biological data, the results of a workflow (often complex), is usually something resembling an Excel spreadsheet with samples as columns and features (genes, microbes, CpG sites, …) as rows. The quantitative information (the numbers) are called “assay” data since they result from running an “assay.” The microbiome is the collection of all microbes (such as bacteria, viruses, fungi, etc.) in the body. When studying these microbes, data is needed, and that’s where assays come in. An assay is a way of measuring the presence and abundance of different types of microbes in a sample. For example, if you want to know how many bacteria of a certain type are in your gut, you can use an assay to measure this. When storing assays, the original data is count-based. However, the original count-based taxonomic abundance tables may undergo different transformations, such as logarithmic, Centered Log-Ratio (CLR), or relative abundance. These are typically stored in <strong>assays</strong>.</p>
<p>The assays slot contains the experimental data as multiple count matrices. The result of assays is a list of matrices.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 1
names(1): relative_abundance</code></pre>
</div>
</div>
<p>Individual assays can be accessed via <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay()</a></code> function. The result is a matrix. And recall that when we subset a matrix, we use <code>[]</code> and specify the <code>[&lt;rows&gt;, &lt;columns&gt;]</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># just the first 5 rows and columns to make the results easier to see</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"relative_abundance"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                   SID31004 SID31009 SID31021 SID31030 SID31071
[Bacteroides] pectinophilus               0  0.35998  0.00000  0.00000        0
[Butyribacterium] methylotrophicum        0  0.00000  0.00000  0.00000        0
[Clostridium] hylemonae                   0  0.00000  0.00000  0.00000        0
[Clostridium] innocuum                    0  0.00371  0.00384  0.01735        0
[Clostridium] leptum                      0  0.00723  0.03135  0.03833        0</code></pre>
</div>
</div>
<p>You can, of course, choose a different number of rows or columns it you like.</p>
</section><section id="sample-information-coldata" class="level3" data-number="24.4.2"><h3 data-number="24.4.2" class="anchored" data-anchor-id="sample-information-coldata">
<span class="header-section-number">24.4.2</span> Sample information: <code>colData</code>
</h3>
<p>colData contains information about the samples used in the study. This information can include details such as the sample ID, the primers used in the analysis, the barcodes associated with the sample (truncated or complete), the type of sample (e.g.&nbsp;soil, fecal, mock) and a description of the sample. You can also add experimental details about samples as you see fit (eg., the sample preparation data, prep kit lot number, or anything else you might want to track).</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 154 rows and 28 columns
           study_name  subject_id   body_site antibiotics_current_use
          &lt;character&gt; &lt;character&gt; &lt;character&gt;             &lt;character&gt;
SID31004   FengQ_2015    SID31004       stool                      no
SID31009   FengQ_2015    SID31009       stool                      no
SID31021   FengQ_2015    SID31021       stool                      no
SID31030   FengQ_2015    SID31030       stool                      no
SID31071   FengQ_2015    SID31071       stool                      no
...               ...         ...         ...                     ...
SID532796  FengQ_2015   SID532796       stool                      no
SID532802  FengQ_2015   SID532802       stool                      no
SID532826  FengQ_2015   SID532826       stool                      no
SID532832  FengQ_2015   SID532832       stool                      no
SID532915  FengQ_2015   SID532915       stool                      no
          study_condition                disease       age age_category
              &lt;character&gt;            &lt;character&gt; &lt;integer&gt;  &lt;character&gt;
SID31004              CRC CRC;fatty_liver;hype..        64        adult
SID31009          control fatty_liver;hyperten..        68       senior
SID31021          control                healthy        60        adult
SID31030          adenoma adenoma;fatty_liver;..        70       senior
SID31071          control            fatty_liver        68       senior
...                   ...                    ...       ...          ...
SID532796         control fatty_liver;hyperten..        73       senior
SID532802         control                healthy        68       senior
SID532826         control T2D;fatty_liver;hype..        78       senior
SID532832         adenoma adenoma;fatty_liver;..        68       senior
SID532915         control                healthy        43        adult
               gender     country non_westernized sequencing_platform
          &lt;character&gt; &lt;character&gt;     &lt;character&gt;         &lt;character&gt;
SID31004         male         AUT              no       IlluminaHiSeq
SID31009         male         AUT              no       IlluminaHiSeq
SID31021       female         AUT              no       IlluminaHiSeq
SID31030         male         AUT              no       IlluminaHiSeq
SID31071         male         AUT              no       IlluminaHiSeq
...               ...         ...             ...                 ...
SID532796      female         AUT              no       IlluminaHiSeq
SID532802        male         AUT              no       IlluminaHiSeq
SID532826        male         AUT              no       IlluminaHiSeq
SID532832      female         AUT              no       IlluminaHiSeq
SID532915      female         AUT              no       IlluminaHiSeq
          DNA_extraction_kit        PMID number_reads number_bases
                 &lt;character&gt; &lt;character&gt;    &lt;integer&gt;    &lt;numeric&gt;
SID31004               MoBio    25758642     40898340   3649611221
SID31009               MoBio    25758642     66107961   6196998053
SID31021               MoBio    25758642     60789126   5708593447
SID31030               MoBio    25758642     50300253   4741158330
SID31071               MoBio    25758642     51945426   4913627034
...                      ...         ...          ...          ...
SID532796              MoBio    25758642     50845712   4754848864
SID532802              MoBio    25758642     41480415   3868696049
SID532826              MoBio    25758642     35346002   3348368467
SID532832              MoBio    25758642     42184599   3951224696
SID532915              MoBio    25758642     51594677   4886998833
          minimum_read_length median_read_length      NCBI_accession
                    &lt;integer&gt;          &lt;integer&gt;         &lt;character&gt;
SID31004                   30                 93 ERR688505;ERR688358
SID31009                   30                 96 ERR688506;ERR688359
SID31021                   30                 96 ERR688507;ERR688360
SID31030                   30                 96 ERR688508;ERR688361
SID31071                   30                 97 ERR688509;ERR688362
...                       ...                ...                 ...
SID532796                  30                 95 ERR710428;ERR710419
SID532802                  30                 96 ERR710429;ERR710420
SID532826                  30                 97 ERR710430;ERR710421
SID532832                  30                 96 ERR710431;ERR710422
SID532915                  30                 96 ERR710432;ERR710423
                         curator       BMI        diet disease_subtype
                     &lt;character&gt; &lt;numeric&gt; &lt;character&gt;     &lt;character&gt;
SID31004  Paolo_Manghi;Marisa_..     29.35  vegetarian       carcinoma
SID31009  Paolo_Manghi;Marisa_..     32.00    omnivore              NA
SID31021  Paolo_Manghi;Marisa_..     22.10    omnivore              NA
SID31030  Paolo_Manghi;Marisa_..     34.11    omnivore advancedadenoma
SID31071  Paolo_Manghi;Marisa_..     23.45    omnivore              NA
...                          ...       ...         ...             ...
SID532796 Paolo_Manghi;Marisa_..     26.56    omnivore              NA
SID532802 Paolo_Manghi;Marisa_..     23.53    omnivore              NA
SID532826 Paolo_Manghi;Marisa_..     31.22    omnivore              NA
SID532832 Paolo_Manghi;Marisa_..     27.55    omnivore advancedadenoma
SID532915 Paolo_Manghi;Marisa_..     22.65    omnivore              NA
                  tnm triglycerides       hdl       ldl     hba1c
          &lt;character&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
SID31004       t1n0m0           172        28        92       5.2
SID31009           NA           101        50       157        NA
SID31021           NA            53        60       122        NA
SID31030           NA            89        74       146        NA
SID31071           NA           258        40       231        NA
...               ...           ...       ...       ...       ...
SID532796          NA           100        61       114       5.4
SID532802          NA            80        67       158       5.8
SID532826          NA           212        32        90       6.9
SID532832          NA           141        51       184       5.6
SID532915          NA            51        80       132       5.1</code></pre>
</div>
</div>
<p>As you can see, there is a lot of information stored in the <code>colData</code> slot, including sample IDs, study conditions, and other metadata associated with the samples. This information is essential for understanding the context of the microbiome data and for performing downstream analyses.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Use sample colData as database of information about your samples.
</div>
</div>
<div class="callout-body-container callout-body">
<p>By keeping a relatively complete colData, you can answer questions about your findings without having to refer back to external Excel spreadsheets or lab notebooks.</p>
</div>
</div>
</section><section id="measured-feature-information-rowdata" class="level3" data-number="24.4.3"><h3 data-number="24.4.3" class="anchored" data-anchor-id="measured-feature-information-rowdata">
<span class="header-section-number">24.4.3</span> Measured feature information: <code>rowData</code>
</h3>
<p>rowData contains data on the features of the analyzed samples. This is particularly important in the microbiome field for storing taxonomic information. This taxonomic information is extremely important for understanding the composition and diversity of the microbiome in each sample analyzed. It enables identification of the different types of microorganisms present in samples. It also allows you to explore the relationships between microbiome composition and various environmental or health factors.</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 7 columns
                                   superkingdom      phylum            class
                                    &lt;character&gt; &lt;character&gt;      &lt;character&gt;
[Bacteroides] pectinophilus            Bacteria  Firmicutes       Clostridia
[Butyribacterium] methylotrophicum     Bacteria  Firmicutes       Clostridia
[Clostridium] hylemonae                Bacteria  Firmicutes       Clostridia
[Clostridium] innocuum                 Bacteria  Firmicutes Erysipelotrichia
[Clostridium] leptum                   Bacteria  Firmicutes       Clostridia
[Clostridium] methylpentosum           Bacteria  Firmicutes       Clostridia
                                                order              family
                                          &lt;character&gt;         &lt;character&gt;
[Bacteroides] pectinophilus             Eubacteriales                  NA
[Butyribacterium] methylotrophicum      Eubacteriales      Clostridiaceae
[Clostridium] hylemonae                 Eubacteriales     Lachnospiraceae
[Clostridium] innocuum             Erysipelotrichales Erysipelotrichaceae
[Clostridium] leptum                    Eubacteriales    Oscillospiraceae
[Clostridium] methylpentosum            Eubacteriales    Oscillospiraceae
                                                    genus
                                              &lt;character&gt;
[Bacteroides] pectinophilus                            NA
[Butyribacterium] methylotrophicum            Clostridium
[Clostridium] hylemonae                 Lachnoclostridium
[Clostridium] innocuum             Erysipelatoclostridium
[Clostridium] leptum                                   NA
[Clostridium] methylpentosum                           NA
                                                  species
                                              &lt;character&gt;
[Bacteroides] pectinophilus        [Bacteroides] pectin..
[Butyribacterium] methylotrophicum [Butyribacterium] me..
[Clostridium] hylemonae            [Clostridium] hylemo..
[Clostridium] innocuum             [Clostridium] innocuum
[Clostridium] leptum                 [Clostridium] leptum
[Clostridium] methylpentosum       [Clostridium] methyl..</code></pre>
</div>
</div>
</section><section id="rowtree" class="level3" data-number="24.4.4"><h3 data-number="24.4.4" class="anchored" data-anchor-id="rowtree">
<span class="header-section-number">24.4.4</span> rowTree</h3>
<p>Phylogenetic trees also play an important role in the microbiome field. The TreeSE class can keep track of features and node relations via two functions, rowTree and rowLinks.</p>
<p>A tree can be accessed via <code><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree()</a></code> as <code>phylo</code> object.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is this tree of which we speak?
</div>
</div>
<div class="callout-body-container callout-body">
<p>The “tree” in a <code>TreeSummarizedExperiment</code> is a phylogenetic tree describing the inferred evolutionary relationships among various biological species or other entities based upon similarities and differences in their physical or genetic characteristics. The tree of life is a phylogenetic tree that shows the evolutionary relationships among all living organisms on Earth.</p>
<p>In the context of microbiome analysis, phylogenetic trees are used to represent the evolutionary relationships between different microbial taxa based on their genetic sequences. These trees can help researchers understand the diversity and relatedness of different microbes in a sample and provide insights into the evolutionary history of the microbial community.</p>
<p>The <code>phylo</code> class in R is used to represent phylogenetic trees and provides functions for working with and visualizing these trees. The <code>ggtree</code> package is a popular package for visualizing phylogenetic trees in R and provides a wide range of options for customizing the appearance of the tree.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 10430 tips and 10429 internal nodes.

Tip labels:
  k__Archaea|p__Candidatus_Micrarchaeota|c__Candidatus_Micrarchaeota_unclassified|o__Candidatus_Micrarchaeota_unclassified|f__Candidatus_Micrarchaeota_unclassified|g__Candidatus_Micrarchaeota_unclassified|s__Candidatus_Micrarchaeota_archaeon_CG1_02_55_22, k__Archaea|p__Archaea_unclassified|c__Archaea_unclassified|o__Archaea_unclassified|f__Archaea_unclassified|g__Archaea_unclassified|s__archaeon_GW2011_AR15, k__Archaea|p__Candidatus_Diapherotrites|c__Candidatus_Diapherotrites_unclassified|o__Candidatus_Diapherotrites_unclassified|f__Candidatus_Diapherotrites_unclassified|g__Candidatus_Diapherotrites_unclassified|s__Candidatus_Diapherotrites_archaeon_CG08_land_8_20_14_0_20_34_12, k__Archaea|p__Archaea_unclassified|c__Archaea_unclassified|o__Archaea_unclassified|f__Archaea_unclassified|g__Archaea_unclassified|s__archaeon_GW2011_AR10, k__Archaea|p__Candidatus_Diapherotrites|c__Candidatus_Diapherotrites_unclassified|o__Candidatus_Diapherotrites_unclassified|f__Candidatus_Diapherotrites_unclassified|g__Candidatus_Diapherotrites_unclassified|s__Candidatus_Diapherotrites_archaeon_CG11_big_fil_rev_8_21_14_0_20_37_9, k__Archaea|p__Euryarchaeota|c__Euryarchaeota_unclassified|o__Euryarchaeota_unclassified|f__Euryarchaeota_unclassified|g__Euryarchaeota_unclassified|s__Euryarchaeota_archaeon_TMED173, ...

Rooted; includes branch length(s).</code></pre>
</div>
</div>
<p>The <code>rowTree</code> slot contains information about the hierarchical structure of the data, such as the relationships between different microbial taxa based on their genetic sequences. This information can be used to explore the evolutionary relationships between different microbes and to visualize the diversity and relatedness of the microbial community in a sample.</p>
<p>We can visualize the phylogenetic tree using the <code>ggtree</code> package in R. The <code>ggtree</code> package provides functions for visualizing phylogenetic trees in a variety of formats, including circular, rectangular, and radial layouts. Here, we’ll use the <code>ggtree</code> package to visualize the phylogenetic tree stored in the <code>rowTree</code> slot of the <code>TreeSummarizedExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/">ggtree</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>See the <a href="https://bioconductor.org/packages/ggtree">ggtree</a> package for more details on visualizing tree structures in R and, for more fun, the <a href="https://yulab-smu.top/treedata-book/">ggtree book</a>.</p>
<p>The <code>rowLink</code> slot contains information about the relationships between the features and the tree structure. This information can be used to link the features in the data to the nodes in the tree and to explore the relationships between the features based on their abundance profiles.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowLinks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LinkDataFrame with 601 rows and 5 columns
                                                  nodeLab nodeLab_alias
                                              &lt;character&gt;   &lt;character&gt;
[Bacteroides] pectinophilus        k__Bacteria|p__Firmi..    alias_3706
[Butyribacterium] methylotrophicum k__Bacteria|p__Firmi..    alias_3347
[Clostridium] hylemonae            k__Bacteria|p__Firmi..    alias_3662
[Clostridium] innocuum             k__Bacteria|p__Firmi..    alias_4453
[Clostridium] leptum               k__Bacteria|p__Firmi..    alias_3525
...                                                   ...           ...
Veillonella sp. T11011-6           k__Bacteria|p__Firmi..    alias_3074
Veillonella tobetsuensis           k__Bacteria|p__Firmi..    alias_3070
Victivallis vadensis               k__Bacteria|p__Lenti..    alias_5062
Weissella cibaria                  k__Bacteria|p__Firmi..    alias_4775
Weissella viridescens              k__Bacteria|p__Firmi..    alias_4773
                                     nodeNum    isLeaf   whichTree
                                   &lt;integer&gt; &lt;logical&gt; &lt;character&gt;
[Bacteroides] pectinophilus             3706      TRUE       phylo
[Butyribacterium] methylotrophicum      3347      TRUE       phylo
[Clostridium] hylemonae                 3662      TRUE       phylo
[Clostridium] innocuum                  4453      TRUE       phylo
[Clostridium] leptum                    3525      TRUE       phylo
...                                      ...       ...         ...
Veillonella sp. T11011-6                3074      TRUE       phylo
Veillonella tobetsuensis                3070      TRUE       phylo
Victivallis vadensis                    5062      TRUE       phylo
Weissella cibaria                       4775      TRUE       phylo
Weissella viridescens                   4773      TRUE       phylo</code></pre>
</div>
</div>
<p>Both <code>rowTree</code> and <code>rowLinks</code> are optional components of the <code>TreeSummarizedExperiment</code> object, but when present, they provide valuable information about the hierarchical structure of the data and the relationships between the features in the data.</p>
</section></section><section id="wrangling-and-subsetting" class="level2" data-number="24.5"><h2 data-number="24.5" class="anchored" data-anchor-id="wrangling-and-subsetting">
<span class="header-section-number">24.5</span> Wrangling and subsetting</h2>
<p>One of the huge advantages of using the <code>SummarizedExperiment</code> classes in R/Bioconductor is their ability to encapsulate all the information necessary to describe the samples (colData), measured features (rowData, rowTree), and the measurements themselves (assays). When we manipulate or subset these objects, all the associated information about samples, features, and assays are also subsetted.</p>
<p>The <code>mia</code> package provides several functions for wrangling and subsetting microbiome data. These functions allow you to filter, transform, and manipulate the data to extract the information you need for analysis.</p>
<section id="subsetting-samples" class="level3" data-number="24.5.1"><h3 data-number="24.5.1" class="anchored" data-anchor-id="subsetting-samples">
<span class="header-section-number">24.5.1</span> Subsetting samples</h3>
<p>You can subset the samples in a <code>TreeSummarizedExperiment</code> object similarly to how you would subset a <code>data.frame</code>. For example, to subset out data based on the age category of the samples, you can use the following code:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 64 68 60 70 68 66</code></pre>
</div>
</div>
<p>Since accessing the colData is such a common way of interacting with a <code>SummarizedExperiment</code>, we can also access the colData as if it were a data.frame.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 64 68 60 70 68 66</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># and get a summary of all patient ages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  43.00   63.00   68.00   66.85   72.00   86.00 </code></pre>
</div>
</div>
<p>Age is a numeric variable. Let’s focus on older patients.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Include all rows, but only columns (samples) who are older than 50</span></span>
<span><span class="va">tse_subset_by_age</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span>, <span class="va">tse</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">tse_subset_by_age</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 601 146 
metadata(1): agglomerated_by_rank
assays(1): relative_abundance
rownames(601): [Bacteroides] pectinophilus [Butyribacterium]
  methylotrophicum ... Weissella cibaria Weissella viridescens
rowData names(7): superkingdom phylum ... genus species
colnames(146): SID31004 SID31009 ... SID532826 SID532832
colData names(28): study_name subject_id ... ldl hba1c
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (601 rows)
rowTree: 1 phylo tree(s) (10430 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>The <code>tse_subset_by_age</code> is a copy of the original <code>tse</code>, but with only the 50+ year old patients. If you want to double-check, you can look at the ages in this new <code>tse_subset_by_age</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">tse_subset_by_age</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  52.00   63.25   68.00   68.02   72.00   86.00 </code></pre>
</div>
</div>
</section><section id="agglomerating-data" class="level3" data-number="24.5.2"><h3 data-number="24.5.2" class="anchored" data-anchor-id="agglomerating-data">
<span class="header-section-number">24.5.2</span> Agglomerating data</h3>
<p>The microbiome features (organisms) that are measured may be measured at different taxonomic levels. For example, the data may be available at the species, genus, family, or phylum level. Agglomerating data is the process of summarizing the data at a higher taxonomic level by combining the abundances of the lower-level taxa. This can be useful for simplifying the data and reducing the dimensionality of the data.</p>
<p>The <code>agglomerateByRank</code> function in the <code>mia</code> package can be used to agglomerate the data at a specified taxonomic rank. For example, to agglomerate the data at the phylum level, you can use the following code:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">new_tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"phylum"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following values are already present in `metadata` and will be
overwritten: 'agglomerated_by_rank'. Consider using the 'name' argument to
specify alternative names.</code></pre>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># observe the new object, particularly the number of features (rows)</span></span>
<span><span class="va">new_tse</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 13 154 
metadata(1): agglomerated_by_rank
assays(1): relative_abundance
rownames(13): Actinobacteria Ascomycota ... Synergistetes
  Verrucomicrobia
rowData names(7): superkingdom phylum ... genus species
colnames(154): SID31004 SID31009 ... SID532832 SID532915
colData names(28): study_name subject_id ... ldl hba1c
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (13 rows)
rowTree: 1 phylo tree(s) (10430 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>The resulting <code>TreeSummarizedExperiment</code> object contains the data aggregated at the phylum level, with the abundances of the lower-level taxa combined to create a summary at the phylum level. This can make the data easier to work with and interpret, especially when dealing with large and complex microbiome datasets.</p>
</section></section><section id="community-indices" class="level2" data-number="24.6"><h2 data-number="24.6" class="anchored" data-anchor-id="community-indices">
<span class="header-section-number">24.6</span> Community indices</h2>
<p>In the field of microbiome ecology several indices to describe samples and community of samples are available. In this vignette we just want to give a very brief introduction.</p>
<section id="alpha-diversity" class="level3" data-number="24.6.1"><h3 data-number="24.6.1" class="anchored" data-anchor-id="alpha-diversity">
<span class="header-section-number">24.6.1</span> Alpha diversity</h3>
<p>diversity in microbiology is measured by several indices:</p>
<ul>
<li>species richness (total number of species)</li>
<li>equitability (distribution of species within a microbiome)</li>
<li>diversity (combination of the two)</li>
</ul>
<p>Functions for calculating alpha and beta diversity indices are available. Using <code>addAlpha</code> multiple diversity indices are calculated by default and results are stored automatically in <code>colData</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why do we care about alpha diversity?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Alpha diversity is a “statistic” that somehow represents how complex or “diverse” a microbiome sample is. You can imagine lots of situations that might significantly alter the diversity of a sample, such as antibiotic use, systemic chemotherapy, autoimmune conditions, dietary changes, etc. When we have two experimental conditions such antibiotic use vs no antibiotic use, the alpha diversity in the two sample groups may differ significantly. That association, if present, could be helpful in understanding biology or even developing a biomarker.</p>
</div>
</div>
<p>In the code below we calculate the Shannon and observed diversity indices.</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addAlpha.html">addAlpha</a></span><span class="op">(</span><span class="va">tse</span>, index <span class="op">=</span> <span class="st">"shannon"</span>, assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addAlpha.html">addAlpha</a></span><span class="op">(</span><span class="va">tse</span>, index <span class="op">=</span> <span class="st">"observed"</span>, assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "study_name"              "subject_id"             
 [3] "body_site"               "antibiotics_current_use"
 [5] "study_condition"         "disease"                
 [7] "age"                     "age_category"           
 [9] "gender"                  "country"                
[11] "non_westernized"         "sequencing_platform"    
[13] "DNA_extraction_kit"      "PMID"                   
[15] "number_reads"            "number_bases"           
[17] "minimum_read_length"     "median_read_length"     
[19] "NCBI_accession"          "curator"                
[21] "BMI"                     "diet"                   
[23] "disease_subtype"         "tnm"                    
[25] "triglycerides"           "hdl"                    
[27] "ldl"                     "hba1c"                  
[29] "shannon"                 "observed"               </code></pre>
</div>
</div>
<p>The <code>observed</code> index is a simple count of the number of species present in a sample, while the <code>shannon</code> index takes into account both the number of species and their relative abundances. These indices provide information about the diversity and richness of the microbial community in each sample.</p>
<p>And we can look at the resulting colData() which now contains the calculated alpha diversity indices.</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 154 rows and 30 columns
           study_name  subject_id   body_site antibiotics_current_use
          &lt;character&gt; &lt;character&gt; &lt;character&gt;             &lt;character&gt;
SID31004   FengQ_2015    SID31004       stool                      no
SID31009   FengQ_2015    SID31009       stool                      no
SID31021   FengQ_2015    SID31021       stool                      no
SID31030   FengQ_2015    SID31030       stool                      no
SID31071   FengQ_2015    SID31071       stool                      no
...               ...         ...         ...                     ...
SID532796  FengQ_2015   SID532796       stool                      no
SID532802  FengQ_2015   SID532802       stool                      no
SID532826  FengQ_2015   SID532826       stool                      no
SID532832  FengQ_2015   SID532832       stool                      no
SID532915  FengQ_2015   SID532915       stool                      no
          study_condition                disease       age age_category
              &lt;character&gt;            &lt;character&gt; &lt;integer&gt;  &lt;character&gt;
SID31004              CRC CRC;fatty_liver;hype..        64        adult
SID31009          control fatty_liver;hyperten..        68       senior
SID31021          control                healthy        60        adult
SID31030          adenoma adenoma;fatty_liver;..        70       senior
SID31071          control            fatty_liver        68       senior
...                   ...                    ...       ...          ...
SID532796         control fatty_liver;hyperten..        73       senior
SID532802         control                healthy        68       senior
SID532826         control T2D;fatty_liver;hype..        78       senior
SID532832         adenoma adenoma;fatty_liver;..        68       senior
SID532915         control                healthy        43        adult
               gender     country non_westernized sequencing_platform
          &lt;character&gt; &lt;character&gt;     &lt;character&gt;         &lt;character&gt;
SID31004         male         AUT              no       IlluminaHiSeq
SID31009         male         AUT              no       IlluminaHiSeq
SID31021       female         AUT              no       IlluminaHiSeq
SID31030         male         AUT              no       IlluminaHiSeq
SID31071         male         AUT              no       IlluminaHiSeq
...               ...         ...             ...                 ...
SID532796      female         AUT              no       IlluminaHiSeq
SID532802        male         AUT              no       IlluminaHiSeq
SID532826        male         AUT              no       IlluminaHiSeq
SID532832      female         AUT              no       IlluminaHiSeq
SID532915      female         AUT              no       IlluminaHiSeq
          DNA_extraction_kit        PMID number_reads number_bases
                 &lt;character&gt; &lt;character&gt;    &lt;integer&gt;    &lt;numeric&gt;
SID31004               MoBio    25758642     40898340   3649611221
SID31009               MoBio    25758642     66107961   6196998053
SID31021               MoBio    25758642     60789126   5708593447
SID31030               MoBio    25758642     50300253   4741158330
SID31071               MoBio    25758642     51945426   4913627034
...                      ...         ...          ...          ...
SID532796              MoBio    25758642     50845712   4754848864
SID532802              MoBio    25758642     41480415   3868696049
SID532826              MoBio    25758642     35346002   3348368467
SID532832              MoBio    25758642     42184599   3951224696
SID532915              MoBio    25758642     51594677   4886998833
          minimum_read_length median_read_length      NCBI_accession
                    &lt;integer&gt;          &lt;integer&gt;         &lt;character&gt;
SID31004                   30                 93 ERR688505;ERR688358
SID31009                   30                 96 ERR688506;ERR688359
SID31021                   30                 96 ERR688507;ERR688360
SID31030                   30                 96 ERR688508;ERR688361
SID31071                   30                 97 ERR688509;ERR688362
...                       ...                ...                 ...
SID532796                  30                 95 ERR710428;ERR710419
SID532802                  30                 96 ERR710429;ERR710420
SID532826                  30                 97 ERR710430;ERR710421
SID532832                  30                 96 ERR710431;ERR710422
SID532915                  30                 96 ERR710432;ERR710423
                         curator       BMI        diet disease_subtype
                     &lt;character&gt; &lt;numeric&gt; &lt;character&gt;     &lt;character&gt;
SID31004  Paolo_Manghi;Marisa_..     29.35  vegetarian       carcinoma
SID31009  Paolo_Manghi;Marisa_..     32.00    omnivore              NA
SID31021  Paolo_Manghi;Marisa_..     22.10    omnivore              NA
SID31030  Paolo_Manghi;Marisa_..     34.11    omnivore advancedadenoma
SID31071  Paolo_Manghi;Marisa_..     23.45    omnivore              NA
...                          ...       ...         ...             ...
SID532796 Paolo_Manghi;Marisa_..     26.56    omnivore              NA
SID532802 Paolo_Manghi;Marisa_..     23.53    omnivore              NA
SID532826 Paolo_Manghi;Marisa_..     31.22    omnivore              NA
SID532832 Paolo_Manghi;Marisa_..     27.55    omnivore advancedadenoma
SID532915 Paolo_Manghi;Marisa_..     22.65    omnivore              NA
                  tnm triglycerides       hdl       ldl     hba1c   shannon
          &lt;character&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
SID31004       t1n0m0           172        28        92       5.2   3.28881
SID31009           NA           101        50       157        NA   3.10400
SID31021           NA            53        60       122        NA   3.37028
SID31030           NA            89        74       146        NA   2.59465
SID31071           NA           258        40       231        NA   3.14486
...               ...           ...       ...       ...       ...       ...
SID532796          NA           100        61       114       5.4   3.65802
SID532802          NA            80        67       158       5.8   3.19414
SID532826          NA           212        32        90       6.9   3.15717
SID532832          NA           141        51       184       5.6   3.21046
SID532915          NA            51        80       132       5.1   3.48644
           observed
          &lt;numeric&gt;
SID31004        100
SID31009        105
SID31021        115
SID31030        105
SID31071        112
...             ...
SID532796       131
SID532802       109
SID532826       101
SID532832        93
SID532915       126</code></pre>
</div>
</div>
<p>The <code>scater</code> package provides some convenient plotting functions that are designed to work with the <code>SummarizedExperiment</code> class.</p>
<div class="cell" data-messasge="false">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: scuttle</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'scater'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ggtree':

    multiplot</code></pre>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>    <span class="st">"observed"</span>,             <span class="co"># the y-axis value that we just calculated</span></span>
<span>    <span class="st">"study_condition"</span>,      <span class="co"># The x-axis</span></span>
<span>    colour_by <span class="op">=</span> <span class="st">"gender"</span>    <span class="co"># Color the points by gender</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="va">Richness</span><span class="op">[</span><span class="va">Observed</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>  <span class="co"># and give a meaningful label</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/plotAlphaDiversity-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="beta-diversity" class="level3" data-number="24.6.2"><h3 data-number="24.6.2" class="anchored" data-anchor-id="beta-diversity">
<span class="header-section-number">24.6.2</span> Beta diversity</h3>
<p>Community similarity refers to the way microorganisms resemble each other in terms of their composition and abundance of different microbial taxa. This can help us understand to what degree different samples resemble each other and finding key information. In microbiome analysis however, it’s more common to measure the dissimilarity/Beta diversity between two samples A and B using the Bray-Curtis measure which is defined as follows:</p>
<p><span class="math display">\[
BC_{ij} = \frac{\sum_{k} |A_{k} - B_{k}|}{\sum_{k} (A_{k} + B_{k})}
\]</span></p>
<p>where <span class="math inline">\(A_{k}\)</span> and <span class="math inline">\(B_{k}\)</span> are the abundances of taxon <span class="math inline">\(k\)</span> in samples A and B, respectively. The Bray-Curtis dissimilarity ranges from 0 (identical communities) to 1 (completely different communities).</p>
<p>The <code>mia</code> package provides functions for calculating beta diversity indices, such as the Bray-Curtis dissimilarity, Jaccard similarity, and UniFrac distance. These indices can be used to compare the microbial communities between different samples and to identify patterns and relationships between samples based on their microbial composition.</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run PCoA on relabundance assay with Bray-Curtis distances</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://vegandevs.github.io/vegan/">vegan</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">runMDS</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>    FUN <span class="op">=</span> <span class="va">vegdist</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"MDS_bray"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code is a bit to unpack. It calculates the Bray-Curtis dissimilarity between samples in the <code>tse</code> object using the <code>vegdist</code> function from the <code>vegan</code> package. The resulting dissimilarity matrix is then used to perform a Principal Coordinate Analysis (PCoA) using the <code>runMDS</code> function from the <code>scater</code> package. The PCoA is a dimensionality reduction technique that projects the high-dimensional Bray-Curtis dissimilarity matrix onto a lower-dimensional space while preserving the pairwise distances between samples.</p>
<p>The resulting <code>TreeSummarizedExperiment</code> object now contains the Bray-Curtis dissimilarity matrix in the <code>MDS_bray</code> slot, which can be used to visualize the relationships between samples based on their microbial composition.</p>
<p>We can visualize the results of the PCoA using the <code>plotReducedDim</code> function from the <code>scater</code> package. This function creates a plot of the reduced dimensions (PCoA) and colors the samples based on a specified variable, such as the study condition</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create ggplot object</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"MDS_bray"</span>, colour_by <span class="op">=</span> <span class="st">"study_condition"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/showPCoA-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can experiment with different variables to color the samples and explore the relationships between samples based on their microbial composition.</p>
</section></section><section id="microbial-composition" class="level2" data-number="24.7"><h2 data-number="24.7" class="anchored" data-anchor-id="microbial-composition">
<span class="header-section-number">24.7</span> Microbial composition</h2>
<p>Let’s now look at the microbial composition of the samples in the <code>tse</code> object. The microbial composition refers to the relative abundances of different microbial taxa in each sample. This information can provide insights into the diversity and structure of the microbial community in each sample and help identify patterns and relationships between samples based on their microbial composition.</p>
<p>The <code>mia</code> package provides functions for visualizing the microbial composition of samples, such as bar plots, heatmaps, and stacked bar plots. These plots can help you explore the relative abundances of different microbial taxa in each sample and identify patterns and relationships between samples based on their microbial composition.</p>
<section id="abundance" class="level3" data-number="24.7.1"><h3 data-number="24.7.1" class="anchored" data-anchor-id="abundance">
<span class="header-section-number">24.7.1</span> Abundance</h3>
<p>We can start by creating a bar plot of the top 40 most abundant taxa in the dataset. This plot shows the relative abundances of the top 40 taxa in each sample, with the taxa sorted by abundance.</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">miaViz</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggraph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'miaViz'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:mia':

    plotNMDS</code></pre>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotAbundanceDensity.html">plotAbundanceDensity</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>    layout <span class="op">=</span> <span class="st">"jitter"</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">40</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html">percent</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in scale_x_log10(label = scales::percent): log-10 transformation
introduced infinite values.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/plotTopTaxa-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also look at the relative abundances of specific taxa in the dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotAbundanceDensity.html">plotAbundanceDensity</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>    layout <span class="op">=</span> <span class="st">"density"</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>    n <span class="op">=</span> <span class="fl">5</span>, colour_by <span class="op">=</span> <span class="st">"study_condition"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in scale_x_log10(): log-10 transformation introduced infinite values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 53 rows containing non-finite outside the scale range
(`stat_density()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="prevalence" class="level3" data-number="24.7.2"><h3 data-number="24.7.2" class="anchored" data-anchor-id="prevalence">
<span class="header-section-number">24.7.2</span> Prevalence</h3>
<p>Prevalence refers to the proportion of samples in which a taxon is present. This information can help identify the most common and rare taxa in the dataset and provide insights into the distribution of different taxa across samples.</p>
<p>Let’s agglomerate the data at the genus level and then plot the prevalence of the top 20 most prevalent genera in the dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following values are already present in `metadata` and will be
overwritten: 'agglomerated_by_rank'. Consider using the 'name' argument to
specify alternative names.</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="va">tse_genus</span>,</span>
<span>    detection <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">100</span>,</span>
<span>    sort <span class="op">=</span> <span class="cn">TRUE</span>, assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span>,</span>
<span>    as.relative <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Bifidobacterium              Dorea   Faecalibacterium Mediterraneibacter 
         0.7857143          0.7662338          0.7337662          0.7272727 
      Anaerostipes            Blautia 
         0.7272727          0.7077922 </code></pre>
</div>
</div>
</section></section><section id="heatmaps" class="level2" data-number="24.8"><h2 data-number="24.8" class="anchored" data-anchor-id="heatmaps">
<span class="header-section-number">24.8</span> Heatmaps</h2>
<p>Heatmaps are a common way to visualize microbiome data and can provide insights into the relative abundances of different microbial taxa in each sample. To keep things simple, we’ll create a heatmap of the relative abundances of the phyla in the dataset.</p>
<p>We first agglomerate the data at the phylum level and then create a heatmap of the relative abundances of the phyla in the dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="va">tse_phylum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"phylum"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following values are already present in `metadata` and will be
overwritten: 'agglomerated_by_rank'. Consider using the 'name' argument to
specify alternative names.</code></pre>
</div>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse_phylum</span>, <span class="st">"relative_abundance"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="310_microbiome_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="further-fun" class="level2" data-number="24.9"><h2 data-number="24.9" class="anchored" data-anchor-id="further-fun">
<span class="header-section-number">24.9</span> Further fun</h2>
<p>The mia package folks have created a <a href="https://microbiome.github.io/OMA/docs/devel/">book on microbome analysis in R</a> that you can explore for more details on microbiome analysis in R. The book covers a wide range of topics related to microbiome analysis, including data loading, preprocessing, visualization, and statistical analysis. It provides detailed explanations and examples of how to work with microbiome data in R and is a valuable resource for anyone interested in learning more about microbiome analysis.</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./geoquery.html" class="pagination-link" aria-label="Accessing and working with public omics data">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Accessing and working with public omics data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./bioc-summarizedexperiment.html" class="pagination-link" aria-label="Introduction to `SummarizedExperiment`">
        <span class="nav-page-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Introduction to <code>SummarizedExperiment</code></span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="./license.html">
<p>License</p>
</a>
  </li>  
</ul>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/seandavi/RBiocBook/edit/main/310_microbiome.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/seandavi/RBiocBook/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/seandavi/RBiocBook/blob/main/310_microbiome.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>