<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Sean Davis">
<meta name="dcterms.date" content="2023-07-11">
<meta name="description" content="A collection of chapters more than a book, but….">
<title>The RBioc Book - 20&nbsp; ATAC-Seq with Bioconductor</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./ranges_and_signals.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./geoquery.html">Bioconductor</a></li><li class="breadcrumb-item"><a href="./atac.html"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">ATAC-Seq with Bioconductor</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The RBioc Book</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/seandavi/RBiocBook" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="./The-RBioc-Book.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./The-RBioc-Book.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
</div>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-1" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-1">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introducing R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r_intro_mechanics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R mechanics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Up and Running with R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages_and_dice.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Packages and more dice</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./data_structures_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview of R Data Structures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./vectors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Vectors</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./matrices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dataframes_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data Frames</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./factors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Factors</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./eda_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exploratory data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dplyr_intro_msleep.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to dplyr: mammal sleep dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eda_and_univariate_brfss.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Case Study: Behavioral Risk Factor Surveillance System</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Statistics and Machine Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./norm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Working with distribution functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./t-stats-and-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">The t-statistic and t-distribution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./kmeans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">K-means clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine_learning_mlr3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Machine Learning 2</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Bioconductor</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./geoquery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing and working with public omics data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bioc-summarizedexperiment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Introduction to <code>SummarizedExperiment</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./eda_with_pca.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">EDA with PCA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ranges_and_signals.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Genomic ranges and features</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./atac.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">ATAC-Seq with Bioconductor</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Appendix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./additional_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Additional resources</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul class="collapse">
<li><a href="#pre-requisites" id="toc-pre-requisites" class="nav-link" data-scroll-target="#pre-requisites">Pre-requisites</a></li>
  <li><a href="#participation" id="toc-participation" class="nav-link" data-scroll-target="#participation">Participation</a></li>
  <li><a href="#r-bioconductor-packages-used" id="toc-r-bioconductor-packages-used" class="nav-link" data-scroll-target="#r-bioconductor-packages-used"><em>R</em> / <em>Bioconductor</em> packages used</a></li>
  <li><a href="#time-outline" id="toc-time-outline" class="nav-link" data-scroll-target="#time-outline">Time outline</a></li>
  <li><a href="#learning-goals" id="toc-learning-goals" class="nav-link" data-scroll-target="#learning-goals">Learning goals</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li>
<a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background"><span class="header-section-number">20.1</span> Background</a>
  <ul class="collapse">
<li><a href="#informatics-overview" id="toc-informatics-overview" class="nav-link" data-scroll-target="#informatics-overview"><span class="header-section-number">20.1.1</span> Informatics overview</a></li>
  <li><a href="#working-with-sequencing-data-in-bioconductor" id="toc-working-with-sequencing-data-in-bioconductor" class="nav-link" data-scroll-target="#working-with-sequencing-data-in-bioconductor"><span class="header-section-number">20.1.2</span> Working with sequencing data in Bioconductor</a></li>
  </ul>
</li>
  <li>
<a href="#data-import-and-quality-control" id="toc-data-import-and-quality-control" class="nav-link" data-scroll-target="#data-import-and-quality-control"><span class="header-section-number">20.2</span> Data import and quality control</a>
  <ul class="collapse">
<li><a href="#coverage" id="toc-coverage" class="nav-link" data-scroll-target="#coverage"><span class="header-section-number">20.2.1</span> Coverage</a></li>
  <li><a href="#fragment-lengths" id="toc-fragment-lengths" class="nav-link" data-scroll-target="#fragment-lengths"><span class="header-section-number">20.2.2</span> Fragment Lengths</a></li>
  </ul>
</li>
  <li><a href="#viewing-data-in-igv" id="toc-viewing-data-in-igv" class="nav-link" data-scroll-target="#viewing-data-in-igv"><span class="header-section-number">20.3</span> Viewing data in IGV</a></li>
  <li>
<a href="#seqence-extraction" id="toc-seqence-extraction" class="nav-link" data-scroll-target="#seqence-extraction"><span class="header-section-number">20.4</span> Seqence extraction</a>
  <ul class="collapse">
<li><a href="#sequences-in-bioconductor" id="toc-sequences-in-bioconductor" class="nav-link" data-scroll-target="#sequences-in-bioconductor"><span class="header-section-number">20.4.1</span> Sequences in Bioconductor</a></li>
  <li><a href="#bsgenomes" id="toc-bsgenomes" class="nav-link" data-scroll-target="#bsgenomes"><span class="header-section-number">20.4.2</span> BSGenome(s)</a></li>
  </ul>
</li>
  <li><a href="#common-peaks" id="toc-common-peaks" class="nav-link" data-scroll-target="#common-peaks"><span class="header-section-number">20.5</span> Common peaks</a></li>
  <li><a href="#additional-work" id="toc-additional-work" class="nav-link" data-scroll-target="#additional-work"><span class="header-section-number">20.6</span> Additional work</a></li>
  <li>
<a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix">Appendix</a>
  <ul class="collapse">
<li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#macs2" id="toc-macs2" class="nav-link" data-scroll-target="#macs2">MACS2</a></li>
  </ul>
</li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/seandavi/RBiocBook/edit/main/atac.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">20</span>&nbsp; <span class="chapter-title">ATAC-Seq with Bioconductor</span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source" data-quarto-source-url="https://github.com/seandavi/RBiocBook/blob/main/atac.qmd"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://seandavi.github.io/">Sean Davis</a> </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://medschool.cuanschutz.edu/">
            University of Colorado<br>Anschutz School of Medicine
            </a>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 11, 2023</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Abstract</div>
    <p>This workshop employs Bioconductor GenomicRanges and GenomicFeatures infrastructure to perform basic import of aligned reads followed by quality control of an ATAC-Seq dataset. It will also examine how to use fragment length (insert size) to isolate chromatin compartments of interest such as putative nucleosome-free regions. Moving on to a multi-sample analysis, peak regions (called separately) will inform similarities between samples. Statistical identification of differential accessibility will provide a second measure of sample differences and serve as potential regions of interest for enrichment analysis and biological hypothesis generation. Data visualization throughout using the Integrated Genome Viewer (IGV) can make findings more intuitive, so students are encouraged to export datasets from R in formats compatible with IGV.</p>
  </div>
</div>

</header><section id="overview" class="level1 unnumbered page-columns page-full"><h1 class="unnumbered">Overview</h1>
<section id="pre-requisites" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="pre-requisites">Pre-requisites</h2>
<p>This workshop assumes:</p>
<ul>
<li>A working and up-to-date version of R</li>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the <a href="http://bioconductor.org/packages/GenomicRanges"><code>GenomicRanges</code></a> package and range manipulations</li>
<li>Familiarity with BAM files and their contents</li>
</ul></section><section id="participation" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="participation">Participation</h2>
<p>After a very brief review of ATAC-Seq and chromatin accessibility, students will work independently to follow this workflow. Additional materials are provided as links at the end of the workshop for those wanting deeper exposure. Additional materials include alignment from FASTQ files and peak calling.</p>
</section><section id="r-bioconductor-packages-used" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li>
<a href="http://bioconductor.org/packages/Rsamtools"><code>Rsamtools</code></a><br>
</li>
<li>
<a href="http://bioconductor.org/packages/GenomicRanges"><code>GenomicRanges</code></a><br>
</li>
<li>
<a href="http://bioconductor.org/packages/GenomicFeatures"><code>GenomicFeatures</code></a><br>
</li>
<li><a href="http://bioconductor.org/packages/GenomicAlignments"><code>GenomicAlignments</code></a></li>
<li><a href="http://bioconductor.org/packages/rtracklayer"><code>rtracklayer</code></a></li>
<li><a href="http://bioconductor.org/packages/heatmaps"><code>heatmaps</code></a></li>
<li><a href="http://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg19.knownGene"><code>TxDb.Hsapiens.UCSC.hg19.knownGene</code></a></li>
</ul></section><section id="time-outline" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="time-outline">Time outline</h2>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Independent work</td>
<td>2-3hr</td>
</tr>
<tr class="odd">
<td>Additional exercises (optional, external)</td>
<td>up to 12 hours</td>
</tr>
</tbody>
</table></section><section id="learning-goals" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="learning-goals">Learning goals</h2>
<ul>
<li>Describe how to import sequence alignments in BAM format into R</li>
<li>Relate fragment size to genomic characteristics such as nucleosome occupancy and open chromatin.</li>
<li>Perform basic alignment manipulations in R to enrich ATAC-seq data for chromatin characteristics.</li>
<li>Gain familiarity with the IGV genome browser and examining data in genomic context.</li>
<li>Visualize summaries of genomic signal using profile plots and heatmaps.</li>
</ul></section><section id="learning-objectives" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="learning-objectives">Learning objectives</h2>
<ul>
<li>Load and save genomic data in BAM and BigWig formats [GenomicAlignments and rtracklayer].</li>
<li>Perform basic QC plots from ATAC-Seq data.</li>
<li>Isolate nucleosome-free and mononucleosome regions from ATAC-seq data.</li>
<li>Install and use IGV to visualize data in genomic context.</li>
<li>Create profile plots using the heatmaps package.</li>
</ul></section><section id="background" class="level2 page-columns page-full" data-number="20.1"><h2 data-number="20.1" class="anchored" data-anchor-id="background">
<span class="header-section-number">20.1</span> Background</h2>
<p>Chromatin accessibility assays measure the extent to which DNA is open and accessible. Such assays now use high throughput sequencing as a quantitative readout. DNAse assays, first using microarrays<span class="citation" data-cites="Crawford2006-nt">(<a href="references.html#ref-Crawford2006-nt" role="doc-biblioref">Crawford, Davis, et al. 2006</a>)</span> and then DNAse-Seq <span class="citation" data-cites="Crawford2006-bk">(<a href="references.html#ref-Crawford2006-bk" role="doc-biblioref">Crawford, Holt, et al. 2006</a>)</span>, requires a larger amount of DNA and is labor-indensive and has been largely supplanted by ATAC-Seq <span class="citation" data-cites="Buenrostro2013-dz">(<a href="references.html#ref-Buenrostro2013-dz" role="doc-biblioref">Buenrostro et al. 2013</a>)</span>.</p>
<p>The Assay for Transposase Accessible Chromatin with high-throughput sequencing (ATAC-seq) method maps chromatin accessibility genome-wide. This method quantifies DNA accessibility with a hyperactive Tn5 transposase that cuts and inserts sequencing adapters into regions of chromatin that are accessible. High throughput sequencing of fragments produced by the process map to regions of increased accessibility, transcription factor binding sites, and nucleosome positioning. The method is both fast and sensitive and can be used as a replacement for DNAse <em>and</em> MNase. A schematic of the protocol is given in Figure @ref(fig:protocol).</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/protocol_6a3270faacfa6d28c2bb23ee8fb1b21b">
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="images/ATAC-Seq_Fig.1.jpg" class="img-fluid figure-img" width="492"></p>
<figcaption class="figure-caption margin-caption">Schematic overview of ATAC-Seq protocol. Figure from Wikipedia.</figcaption></figure>
</div>
</div>
</div>
<p>An early review of chromatin accessibility assays <span class="citation" data-cites="Tsompana2014-yh">(<a href="references.html#ref-Tsompana2014-yh" role="doc-biblioref">Tsompana and Buck 2014</a>)</span> compares the use cases, pros and cons, and expected signals from each of the most common approaches (Figure @ref(fig:chromatinAssays)).</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/chromatinAssays_56d36045a261b1ec8ad92d29fe667b56">
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="images/chromatin_accessibility_buck.png" class="img-fluid figure-img" width="808"></p>
<figcaption class="figure-caption margin-caption">Chromatin accessibility methods, compared. Representative DNA fragments generated by each assay are shown, with end locations within chromatin defined by colored arrows. Bar diagrams represent data signal obtained from each assay across the entire region. The footprint created by a transcription factor (TF) is shown for ATAC-seq and DNase-seq experiments.</figcaption></figure>
</div>
</div>
</div>
<p>The first manuscript describing ATAC-Seq protocol and findings outlined how ATAC-Seq data “line up” with other datatypes such as ChIP-seq and DNAse-seq (Figure @ref(fig:greenleaf)). They also highlight how fragment length correlates with specific genomic regions and characteristics <span class="citation" data-cites="Buenrostro2013-dz">(<a href="references.html#ref-Buenrostro2013-dz" role="doc-biblioref">Buenrostro et al. 2013, fig. 3</a>)</span>.</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/greenleaf_8fcbc06b471f9d02de3483c345591b0f">
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="images/greenleaf_chip_dnase_atac.png" class="img-fluid figure-img" width="836"></p>
<figcaption class="figure-caption margin-caption">Multimodal chromatin comparisons. From <span class="citation" data-cites="Buenrostro2013-dz">(<a href="references.html#ref-Buenrostro2013-dz" role="doc-biblioref">Buenrostro et al. 2013</a>)</span>, Figure 4. (a) CTCF footprints observed in ATAC-seq and DNase-seq data, at a specific locus on chr1. (b) Aggregate ATAC-seq footprint for CTCF (motif shown) generated over binding sites within the genome (c) CTCF predicted binding probability inferred from ATAC-seq data, position weight matrix (PWM) scores for the CTCF motif, and evolutionary conservation (PhyloP). Right-most column is the CTCF ChIP-seq data (ENCODE) for this GM12878 cell line, demonstrating high concordance with predicted binding probability.</figcaption></figure>
</div>
</div>
</div>
<p>Buenrostro et al.&nbsp;provide a detailed protocol for performing ATAC-Seq and quality control of results <span class="citation" data-cites="Buenrostro2015-dz">(<a href="references.html#ref-Buenrostro2015-dz" role="doc-biblioref">Buenrostro et al. 2015</a>)</span>. Updated and modified protocols that improve on signal-to-noise and reduce input DNA requirements have been described.</p>
<section id="informatics-overview" class="level3 page-columns page-full" data-number="20.1.1"><h3 data-number="20.1.1" class="anchored" data-anchor-id="informatics-overview">
<span class="header-section-number">20.1.1</span> Informatics overview</h3>
<p>ATAC-Seq protocols typically utilize paired-end sequencing protocols. The reads are aligned to the respective <em>genome</em> using <code>bowtie2</code>, <code>BWA</code>, or other short-read aligner. The result, after appropriate manipulation, often using <code>samtools</code>, results in a BAM file. Among other details, the BAM format includes columns for:</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/bamshot_1a0756827adddbe0afb81387e2724806">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span><span class="st">'images/bam_shot.png'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="images/bam_shot.png" class="img-fluid figure-img" width="1126"></p>
<figcaption class="figure-caption margin-caption">A BAM file in text form. The output of <code>samtools view</code> is the text format of the BAM file (called SAM format). Bioconductor and many other tools use BAM files for input. Note that BAM files also often include an index <code>.bai</code> file that enables random access into the file; one can read just a genomic region without having to read the entire file.</figcaption></figure>
</div>
</div>
</div>
<ul>
<li>sequence name (<code>chr1</code>)</li>
<li>start position (integer)</li>
<li>a <em>CIGAR</em> string that describes the alignment in a compact form</li>
<li>the sequence to which the pair aligns</li>
<li>the position to which the pair aligns</li>
<li>a bit flag field that describes multiple characteristics of the alignment</li>
<li>the sequence and quality string of the read</li>
<li>additional tags that tend to be aligner-specific</li>
</ul>
<p>Duplicate fragments (those with the <em>same</em> start and end position of other reads) are marked and likely discarded. Reads that fail to align “properly” are also often excluded from analysis. It is worth noting that most software packages allow simple “marking” of such reads and that there is usually no need to create a special BAM file before proceeding with downstream work.</p>
<p>After alignment and BAM processing, the workflow can switch to <em>Bioconductor</em>.</p>
</section><section id="working-with-sequencing-data-in-bioconductor" class="level3" data-number="20.1.2"><h3 data-number="20.1.2" class="anchored" data-anchor-id="working-with-sequencing-data-in-bioconductor">
<span class="header-section-number">20.1.2</span> Working with sequencing data in Bioconductor</h3>
<p>The <em>Bioconductor</em> project includes several infrastructure packages for dealing with ranges (sequence name, start, end, +/- strand) on sequences <span class="citation" data-cites="Lawrence2013-he">(<a href="references.html#ref-Lawrence2013-he" role="doc-biblioref">Lawrence et al. 2013</a>)</span> as well as capabilities with working with Fastq files directly <span class="citation" data-cites="morgan2016rsamtools">(<a href="references.html#ref-morgan2016rsamtools" role="doc-biblioref">Morgan et al. 2016</a>)</span>.</p>
<table class="table">
<caption>Commonly used Bioconductor and their high-level use cases.</caption>
<thead><tr class="header">
<th>Package</th>
<th>Use cases</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://bioconductor.org/packages/Rsamtools"><code>Rsamtools</code></a></td>
<td>low level access to FASTQ, VCF, SAM, BAM, BCF formats</td>
</tr>
<tr class="even">
<td><a href="http://bioconductor.org/packages/GenomicRanges"><code>GenomicRanges</code></a></td>
<td>Container and methods for handling genomic reagions</td>
</tr>
<tr class="odd">
<td><a href="http://bioconductor.org/packages/GenomicFeatures"><code>GenomicFeatures</code></a></td>
<td>Work with transcript databases, gff, gtf and BED formats</td>
</tr>
<tr class="even">
<td><a href="http://bioconductor.org/packages/GenomicAlignments"><code>GenomicAlignments</code></a></td>
<td>Reader for BAM format</td>
</tr>
<tr class="odd">
<td><a href="http://bioconductor.org/packages/rtracklayer"><code>rtracklayer</code></a></td>
<td>import and export multiple UCSC file formats including BigWig and Bed</td>
</tr>
</tbody>
</table>
<p>As noted in the previous section, the output of an ATAC-Seq experiment is a BAM file. As paired-end sequencing is a commonly-applied approach for ATAC-Seq, the <code>readGAlignmentPairs</code> function is the appropriate method to use.</p>
</section></section><section id="data-import-and-quality-control" class="level2 page-columns page-full" data-number="20.2"><h2 data-number="20.2" class="anchored" data-anchor-id="data-import-and-quality-control">
<span class="header-section-number">20.2</span> Data import and quality control</h2>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-2_89fc9a8651bfc07d891912550a4381f2">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicAlignments">GenomicAlignments</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Reading a paired-end BAM file looks a bit complicated, but the following code will:</p>
<ol type="1">
<li>Read the included BAM file.</li>
<li>Include read pairs only (<code>isPaired = TRUE</code>)</li>
<li>Include properly paired reads (<code>isProperPair = TRUE</code>)</li>
<li>Include reads with mapping quality &gt;= 1</li>
<li>Add a couple of additional fields, <code>mapq</code> (mapping quality) and <code>isize</code> (insert size) to the default fields.</li>
</ol>
<div class="cell" data-hash="atac_cache/html/readGreenleaf_bc23e7fd6bdc1b668803b8ae9211d3a6">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">greenleaf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html">readGAlignmentPairs</a></span><span class="op">(</span></span>
<span>  <span class="st">'https://github.com/seandavi/AtacSeqWorkshop/raw/master/inst/extdata/Sorted_ATAC_21_22.bam'</span>,</span>
<span>  param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">ScanBamParam</a></span><span class="op">(</span></span>
<span>    mapqFilter <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    flag <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">scanBamFlag</a></span><span class="op">(</span></span>
<span>      isPaired <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>      isProperPair <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>    what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mapq"</span>, <span class="st">"isize"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Exercise</em>: What is the class of <code>greenleaf</code>? <em>Exercise</em>: Use the <code><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">GenomicAlignments::first()</a></code> accessor to get the first read of the pair as a <code>GAlignments</code> object. Save the result as a variable called <code>gl_first_read</code>. Use the <code>mcols</code> accessor to find the “metadata columns” of <code>gl_first_read</code>. <em>Exercise</em>: How many read pairs map to each chromosome?</p>
<p>We can make plot of the number of reads mapping to each chromosome.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-3_62fb9d551a1d0069cbeb8fe111d7611b">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">chromCounts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>chromosome<span class="op">=</span><span class="va">Var1</span>, count <span class="op">=</span> <span class="va">Freq</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To keep things small, the example BAM file includes only chromosomes 21 and 22.</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/plotChromCounts_47939c6179931f6d1d4499257bbc7e9e">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">chromCounts</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">chromosome</span>, y<span class="op">=</span><span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">'identity'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/plotChromCounts-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Reads per chromosome. In our example data, we are using only chromosomes 21 and 22.</figcaption></figure>
</div>
</div>
</div>
<p>Normalizing by the chromosome length can yield the reads per megabase which should crudely be similar across all chromosomes.</p>
<div class="cell" data-hash="atac_cache/html/chromReadProps_a230f1b3760b9e6be5c50ae52d631e09">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">chromCounts</span> <span class="op">=</span> <span class="va">chromCounts</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>readsPerMb <span class="op">=</span> <span class="op">(</span><span class="va">count</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqlengths</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">/</span><span class="fl">1e6</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And show a plot. For two chromosomes, this is a little underwhelming.</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/plotReadProps_5eabc90904fffc7555f070a4280d6933">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">chromCounts</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">chromosome</span>, y<span class="op">=</span><span class="va">readsPerMb</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">'identity'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/plotReadProps-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Read counts normalized by chromosome length. This is not a particularly important plot, but it can be useful to see the relative contribution of each chromosome given its length.</figcaption></figure>
</div>
</div>
</div>
<section id="coverage" class="level3" data-number="20.2.1"><h3 data-number="20.2.1" class="anchored" data-anchor-id="coverage">
<span class="header-section-number">20.2.1</span> Coverage</h3>
<p>The <code>coverage</code> method for genomic ranges calculates, for each base, the number of overlapping features. In the case of a BAM file from ATAC-Seq converted to a GAlignmentPairs object, the coverage gives us an idea of the extent to which reads pile up to form peaks.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-4_e9b69be97ed97ce0649d738414e6f336">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cvg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">cvg</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SimpleRleList"
attr(,"package")
[1] "IRanges"</code></pre>
</div>
</div>
<p>The coverage is returned as a <code>SimpleRleList</code> object. Using <code>names</code> can get us the names of the elements of the list.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-5_9c14811ede53ecb8ded68fb97730b44c">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cvg</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "chr1"  "chr2"  "chr3"  "chr4"  "chr5"  "chr6"  "chr7"  "chr8"  "chr9" 
[10] "chr10" "chr11" "chr12" "chr13" "chr14" "chr15" "chr16" "chr17" "chr18"
[19] "chr19" "chr20" "chr21" "chr22" "chrX"  "chrY"  "chrM" </code></pre>
</div>
</div>
<p>There is a name for each chromosome. Looking at the <code>chr21</code> entry:</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-6_820f609dbaf84e2dfd5e873b805d8304">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cvg</span><span class="op">$</span><span class="va">chr21</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>integer-Rle of length 48129895 with 397462 runs
  Lengths: 9411376      50      11      50 ...      36      14      28   10806
  Values :       0       2       0       2 ...       1       2       1       0</code></pre>
</div>
</div>
<p>we see that each chromosome is represented as an <code>Rle</code>, short for run-length-encoding. Simply put, since along the chromosome there are many repeated values, we can recode the long vector as a set of (length: value) pairs. For example, if the first 9,410,000 base pairs have <code>0</code> coverage, we encode that as (9,410,000: 0). Doing that across the chromosome can very significantly reduce the memory use for genomic coverage.</p>
<p>The following little function, <code>plotCvgHistByChrom</code> can plot a histogram of the coverage for a chromosome.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-7_baad0df403d555758852f756ab636a29">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">plotCvgHistByChrom</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">cvg</span>, <span class="va">chromosome</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span>  <span class="va">cvgcounts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="va">cvg</span><span class="op">[[</span><span class="va">chromosome</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">cvgcounts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">cvgcounts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">cvgcounts</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Coverage'</span>, <span class="st">'Count'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cvgcounts</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Coverage</span>, y<span class="op">=</span><span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"Chromosome"</span>,<span class="va">chromosome</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>span<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="op">}</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'chr21'</span>, <span class="st">'chr22'</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu">plotCvgHistByChrom</span><span class="op">(</span><span class="va">cvg</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="atac_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="atac_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="fragment-lengths" class="level3 page-columns page-full" data-number="20.2.2"><h3 data-number="20.2.2" class="anchored" data-anchor-id="fragment-lengths">
<span class="header-section-number">20.2.2</span> Fragment Lengths</h3>
<p>The first ATAC-Seq manuscript <span class="citation" data-cites="Buenrostro2013-dz">(<a href="references.html#ref-Buenrostro2013-dz" role="doc-biblioref">Buenrostro et al. 2013</a>)</span> highlighted the relationship between fragment length and nucleosomes (see Figure @ref{fig:flgreenleaf}).</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/flgreenleaf_b226048bc3766d117a82464a3f8e6a8b">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span><span class="st">'images/greenleaf_atac.png'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="images/greenleaf_atac.png" class="img-fluid figure-img" width="525"></p>
<figcaption class="figure-caption margin-caption">Relationship between fragment length and nucleosome number.</figcaption></figure>
</div>
</div>
</div>
<p>Remember that we loaded the example BAM file with insert sizes (<code>isize</code>). We can use that “column” to examine the fragment lengths (another name for insert size). Also, note that the insert size for the <code>first</code> read and the <code>second</code> are the same (absolute value). Here, we will use <code>first</code>.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-8_6628a3a45ecf068e4f2fd1ffb14bf6cc">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">isize</span><span class="op">)</span></span>
<span><span class="va">fraglengths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">isize</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can plot the fragment length density (histogram) using the <code>density</code> function.</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/fraglength_edaff8d326416a865732be54b1d610c2">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">fraglengths</span>, bw<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/fraglength-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Fragment length histogram.</figcaption></figure>
</div>
</div>
</div>
<p><em>Exercise</em>: Adjust the <code>xlim</code>, <code>bw</code>, and try <code>log="y"</code> in the plot to highlight features present in figure <span class="math inline">\(\ref{fig:flgreenleaf}\)</span>.</p>
<p>And for fun, the ggplot2 version:</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-9_1a8b878ac17389e5c83ab3d41ba71eb2">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">fragLenPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="va">fraglengths</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>InsertSize <span class="op">=</span> <span class="va">fraglengths</span>, </span>
<span>    Count <span class="op">=</span> <span class="va">Freq</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>InsertSize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html">as.vector</a></span><span class="op">(</span><span class="va">InsertSize</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    Count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html">as.vector</a></span><span class="op">(</span><span class="va">Count</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">InsertSize</span>, y <span class="op">=</span> <span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">fragLenPlot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">lims</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">250</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="atac_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Knowing that the nucleosome-free regions will have insert sizes shorter than one nucleosome, we can isolate the read pairs that have that characteristic.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-10_079f798167afc6bbd45c887c4495e3cc">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gl_nf</span> <span class="op">=</span> <span class="va">greenleaf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">isize</span><span class="op">&lt;</span><span class="fl">100</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And the mononucleosome reads will be between 187 and 250 base pairs for insert size/fragment length.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-11_7908509166a10e59e0ac6bc395333dba">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gl_mn</span> <span class="op">=</span> <span class="va">greenleaf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">isize</span><span class="op">&gt;</span><span class="fl">187</span> <span class="op">&amp;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu">GenomicAlignments</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Pairs-class.html">first</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">isize</span><span class="op">&lt;</span><span class="fl">250</span>  <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we expect nucleosome-free reads to be enriched near the TSS while mononucleosome reads should not be. We will use the <a href="http://bioconductor.org/packages/heatmaps"><code>heatmaps</code></a> package to take a look at these two sets of reads with respect to the tss of the human genome.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-12_cc566b31852fe2d97a778f4df299ea8d">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span><span class="va">proms</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">promoters</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span>, <span class="fl">250</span>, <span class="fl">250</span><span class="op">)</span></span>
<span><span class="va">seqs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'chr21'</span>,<span class="st">'chr22'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqlevels</a></span><span class="op">(</span><span class="va">proms</span>, pruning.mode<span class="op">=</span><span class="st">'coarse'</span><span class="op">)</span> <span class="op">=</span> <span class="va">seqs</span> <span class="co"># only chromosome 21 and 22</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Take a look at the <a href="http://bioconductor.org/packages/heatmaps"><code>heatmaps</code></a> package vignette to learn more about the heatmaps package capabilities.</p>
<div class="cell page-columns page-full" data-hash="atac_cache/html/unnamed-chunk-13_6a0910a8ed96f02a7b6c920c93be3a95">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">heatmaps</span><span class="op">)</span></span>
<span><span class="va">gl_nf_hm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/CoverageHeatmap.html">CoverageHeatmap</a></span><span class="op">(</span><span class="va">proms</span>,<span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">gl_nf</span><span class="op">)</span>,coords<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">250</span>,<span class="fl">250</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/label.html">label</a></span><span class="op">(</span><span class="va">gl_nf_hm</span><span class="op">)</span> <span class="op">=</span> <span class="st">"NucFree"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/scale.html">scale</a></span><span class="op">(</span><span class="va">gl_nf_hm</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/plotHeatmapMeta.html">plotHeatmapMeta</a></span><span class="op">(</span><span class="va">gl_nf_hm</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Enrichment of nucleosome free reads just upstream of the TSS.</figcaption></figure>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-hash="atac_cache/html/unnamed-chunk-14_984931bd3fa2d711220ec930301b8c15">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gl_mn_hm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/CoverageHeatmap.html">CoverageHeatmap</a></span><span class="op">(</span><span class="va">proms</span>,<span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">gl_mn</span><span class="op">)</span>,coords<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">250</span>,<span class="fl">250</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/label.html">label</a></span><span class="op">(</span><span class="va">gl_mn_hm</span><span class="op">)</span> <span class="op">=</span> <span class="st">"MonoNuc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/scale.html">scale</a></span><span class="op">(</span><span class="va">gl_mn_hm</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/plotHeatmapMeta.html">plotHeatmapMeta</a></span><span class="op">(</span><span class="va">gl_mn_hm</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Depletion of nucleosome free reads just upstream of the TSS.</figcaption></figure>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-hash="atac_cache/html/unnamed-chunk-15_d5ca800337ac4b27c8c1b9df334b694e">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/heatmaps/man/plotHeatmapList.html">plotHeatmapList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">gl_mn_hm</span>, <span class="va">gl_nf_hm</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full"><p><img src="atac_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption margin-caption">Comparison of signals at TSS. Mononucleosome data on the left, nucleosome-free on the right.</figcaption></figure>
</div>
</div>
</div>
</section></section><section id="viewing-data-in-igv" class="level2" data-number="20.3"><h2 data-number="20.3" class="anchored" data-anchor-id="viewing-data-in-igv">
<span class="header-section-number">20.3</span> Viewing data in IGV</h2>
<p>Install IGV from <a href="https://software.broadinstitute.org/software/igv/download">here</a>.</p>
<p>We export the greenleaf data as a BigWig file.</p>
<div class="cell" data-hash="atac_cache/html/exportBW_0c7b6f7fca8dc78e0aa1f2c6c6c3e571">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html">export.bw</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">greenleaf</span><span class="op">)</span>,<span class="st">'greenleaf.bw'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Exercise</em>: In IGV, choose <code>hg19</code>. Then, load the <code>greenleaf.bw</code> file and explore chromosomes 21 and 22. <em>Exercise</em>: Export the nucleosome-free portion of the data as a BigWig file and examine that in IGV. Where do you expect to see the strongest signals?</p>
</section><section id="seqence-extraction" class="level2" data-number="20.4"><h2 data-number="20.4" class="anchored" data-anchor-id="seqence-extraction">
<span class="header-section-number">20.4</span> Seqence extraction</h2>
<section id="sequences-in-bioconductor" class="level3" data-number="20.4.1"><h3 data-number="20.4.1" class="anchored" data-anchor-id="sequences-in-bioconductor">
<span class="header-section-number">20.4.1</span> Sequences in Bioconductor</h3>
<p>The BioStrings package provides a number of functions for working with sequences.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-16_57b8a0db0f59c68acbcf443904224a0e">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can create a DNAString object from a string of nucleotides.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-17_7e7fbddb05ac90b86d665830561e14c7">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dna</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span><span class="op">(</span><span class="st">'gcgctgctggatgcgaccgcgcatgcgagcgcgacctatccggaa'</span><span class="op">)</span></span>
<span><span class="va">dna</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>45-letter DNAString object
seq: GCGCTGCTGGATGCGACCGCGCATGCGAGCGCGACCTATCCGGAA</code></pre>
</div>
</div>
<p>You may not have known that DNA has a convention for including multiple possible bases. The entire possible alphabet for DNA is given by:</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-18_78ae5455f41706a006b6d13ffec0191a">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XString-class.html">alphabet</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A" "C" "G" "T" "M" "R" "W" "S" "Y" "K" "V" "H" "D" "B" "N" "-" "+" "."</code></pre>
</div>
</div>
<p>Try to create a DNAString object from a string of amino acids, for example. What happens?</p>
<p>We can reverse complement the sequence.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-19_c5f1dc78a640ee3e84a7bc1f553e923d">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html">reverseComplement</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>45-letter DNAString object
seq: TTCCGGATAGGTCGCGCTCGCATGCGCGGTCGCATCCAGCAGCGC</code></pre>
</div>
</div>
<p>We can translate the sequence into amino acids.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-20_e7f1df68cc0ce46757ecf8ec996097fa">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/translate.html">translate</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>15-letter AAString object
seq: ALLDATAHASATYPE</code></pre>
</div>
</div>
<p>Subsetting a DNAString object returns another DNAString object.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-21_e27b6d8dea2af17295b4e4c96d69b867">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dna</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>10-letter DNAString object
seq: GCGCTGCTGG</code></pre>
</div>
</div>
<p>We can extract the nucleotide frequencies (or amino acid frequencies) from a DNAStringSet object.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-22_6bfc79e09e9ae4f5fa4dd8c7ee59dec8">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">alphabetFrequency</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> A  C  G  T  M  R  W  S  Y  K  V  H  D  B  N  -  +  . 
 8 15 16  6  0  0  0  0  0  0  0  0  0  0  0  0  0  0 </code></pre>
</div>
</div>
</section><section id="bsgenomes" class="level3" data-number="20.4.2"><h3 data-number="20.4.2" class="anchored" data-anchor-id="bsgenomes">
<span class="header-section-number">20.4.2</span> BSGenome(s)</h3>
<p>The <a href="https://bioconductor.org/packages/BSgenome">BSgenome</a> package provides functionality to interact with genome sequences that come pre-packaged.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-23_7579a7f1726b684dd84d37367dd948e2">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BSgenome">BSgenome</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The names of available genomes are below. They are all named <code>BSgenome.&lt;species&gt;.&lt;source&gt;.&lt;assembly&gt;</code> as in <code>BSgenome.Hsapiens.UCSC.hg19</code>, <code>BSgenome.Mmusculus.UCSC.mm10</code>, etc.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-24_2b29c05249b0627a4d1790a61d13303f">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html">available.genomes</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "BSgenome.Alyrata.JGI.v1"                           
  [2] "BSgenome.Amellifera.BeeBase.assembly4"             
  [3] "BSgenome.Amellifera.NCBI.AmelHAv3.1"               
  [4] "BSgenome.Amellifera.UCSC.apiMel2"                  
  [5] "BSgenome.Amellifera.UCSC.apiMel2.masked"           
  [6] "BSgenome.Aofficinalis.NCBI.V1"                     
  [7] "BSgenome.Athaliana.TAIR.04232008"                  
  [8] "BSgenome.Athaliana.TAIR.TAIR9"                     
  [9] "BSgenome.Btaurus.UCSC.bosTau3"                     
 [10] "BSgenome.Btaurus.UCSC.bosTau3.masked"              
 [11] "BSgenome.Btaurus.UCSC.bosTau4"                     
 [12] "BSgenome.Btaurus.UCSC.bosTau4.masked"              
 [13] "BSgenome.Btaurus.UCSC.bosTau6"                     
 [14] "BSgenome.Btaurus.UCSC.bosTau6.masked"              
 [15] "BSgenome.Btaurus.UCSC.bosTau8"                     
 [16] "BSgenome.Btaurus.UCSC.bosTau9"                     
 [17] "BSgenome.Btaurus.UCSC.bosTau9.masked"              
 [18] "BSgenome.Carietinum.NCBI.v1"                       
 [19] "BSgenome.Celegans.UCSC.ce10"                       
 [20] "BSgenome.Celegans.UCSC.ce11"                       
 [21] "BSgenome.Celegans.UCSC.ce2"                        
 [22] "BSgenome.Celegans.UCSC.ce6"                        
 [23] "BSgenome.Cfamiliaris.UCSC.canFam2"                 
 [24] "BSgenome.Cfamiliaris.UCSC.canFam2.masked"          
 [25] "BSgenome.Cfamiliaris.UCSC.canFam3"                 
 [26] "BSgenome.Cfamiliaris.UCSC.canFam3.masked"          
 [27] "BSgenome.Cjacchus.UCSC.calJac3"                    
 [28] "BSgenome.Cjacchus.UCSC.calJac4"                    
 [29] "BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1"
 [30] "BSgenome.Creinhardtii.JGI.v5.6"                    
 [31] "BSgenome.Dmelanogaster.UCSC.dm2"                   
 [32] "BSgenome.Dmelanogaster.UCSC.dm2.masked"            
 [33] "BSgenome.Dmelanogaster.UCSC.dm3"                   
 [34] "BSgenome.Dmelanogaster.UCSC.dm3.masked"            
 [35] "BSgenome.Dmelanogaster.UCSC.dm6"                   
 [36] "BSgenome.Drerio.UCSC.danRer10"                     
 [37] "BSgenome.Drerio.UCSC.danRer11"                     
 [38] "BSgenome.Drerio.UCSC.danRer5"                      
 [39] "BSgenome.Drerio.UCSC.danRer5.masked"               
 [40] "BSgenome.Drerio.UCSC.danRer6"                      
 [41] "BSgenome.Drerio.UCSC.danRer6.masked"               
 [42] "BSgenome.Drerio.UCSC.danRer7"                      
 [43] "BSgenome.Drerio.UCSC.danRer7.masked"               
 [44] "BSgenome.Dvirilis.Ensembl.dvircaf1"                
 [45] "BSgenome.Ecoli.NCBI.20080805"                      
 [46] "BSgenome.Gaculeatus.UCSC.gasAcu1"                  
 [47] "BSgenome.Gaculeatus.UCSC.gasAcu1.masked"           
 [48] "BSgenome.Ggallus.UCSC.galGal3"                     
 [49] "BSgenome.Ggallus.UCSC.galGal3.masked"              
 [50] "BSgenome.Ggallus.UCSC.galGal4"                     
 [51] "BSgenome.Ggallus.UCSC.galGal4.masked"              
 [52] "BSgenome.Ggallus.UCSC.galGal5"                     
 [53] "BSgenome.Ggallus.UCSC.galGal6"                     
 [54] "BSgenome.Gmax.NCBI.Gmv40"                          
 [55] "BSgenome.Hsapiens.1000genomes.hs37d5"              
 [56] "BSgenome.Hsapiens.NCBI.GRCh38"                     
 [57] "BSgenome.Hsapiens.NCBI.T2T.CHM13v2.0"              
 [58] "BSgenome.Hsapiens.UCSC.hg17"                       
 [59] "BSgenome.Hsapiens.UCSC.hg17.masked"                
 [60] "BSgenome.Hsapiens.UCSC.hg18"                       
 [61] "BSgenome.Hsapiens.UCSC.hg18.masked"                
 [62] "BSgenome.Hsapiens.UCSC.hg19"                       
 [63] "BSgenome.Hsapiens.UCSC.hg19.masked"                
 [64] "BSgenome.Hsapiens.UCSC.hg38"                       
 [65] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major"        
 [66] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor"        
 [67] "BSgenome.Hsapiens.UCSC.hg38.masked"                
 [68] "BSgenome.Hsapiens.UCSC.hs1"                        
 [69] "BSgenome.Mdomestica.UCSC.monDom5"                  
 [70] "BSgenome.Mfascicularis.NCBI.5.0"                   
 [71] "BSgenome.Mfascicularis.NCBI.6.0"                   
 [72] "BSgenome.Mfuro.UCSC.musFur1"                       
 [73] "BSgenome.Mmulatta.UCSC.rheMac10"                   
 [74] "BSgenome.Mmulatta.UCSC.rheMac2"                    
 [75] "BSgenome.Mmulatta.UCSC.rheMac2.masked"             
 [76] "BSgenome.Mmulatta.UCSC.rheMac3"                    
 [77] "BSgenome.Mmulatta.UCSC.rheMac3.masked"             
 [78] "BSgenome.Mmulatta.UCSC.rheMac8"                    
 [79] "BSgenome.Mmusculus.UCSC.mm10"                      
 [80] "BSgenome.Mmusculus.UCSC.mm10.masked"               
 [81] "BSgenome.Mmusculus.UCSC.mm39"                      
 [82] "BSgenome.Mmusculus.UCSC.mm8"                       
 [83] "BSgenome.Mmusculus.UCSC.mm8.masked"                
 [84] "BSgenome.Mmusculus.UCSC.mm9"                       
 [85] "BSgenome.Mmusculus.UCSC.mm9.masked"                
 [86] "BSgenome.Osativa.MSU.MSU7"                         
 [87] "BSgenome.Ppaniscus.UCSC.panPan1"                   
 [88] "BSgenome.Ppaniscus.UCSC.panPan2"                   
 [89] "BSgenome.Ptroglodytes.UCSC.panTro2"                
 [90] "BSgenome.Ptroglodytes.UCSC.panTro2.masked"         
 [91] "BSgenome.Ptroglodytes.UCSC.panTro3"                
 [92] "BSgenome.Ptroglodytes.UCSC.panTro3.masked"         
 [93] "BSgenome.Ptroglodytes.UCSC.panTro5"                
 [94] "BSgenome.Ptroglodytes.UCSC.panTro6"                
 [95] "BSgenome.Rnorvegicus.UCSC.rn4"                     
 [96] "BSgenome.Rnorvegicus.UCSC.rn4.masked"              
 [97] "BSgenome.Rnorvegicus.UCSC.rn5"                     
 [98] "BSgenome.Rnorvegicus.UCSC.rn5.masked"              
 [99] "BSgenome.Rnorvegicus.UCSC.rn6"                     
[100] "BSgenome.Rnorvegicus.UCSC.rn7"                     
[101] "BSgenome.Scerevisiae.UCSC.sacCer1"                 
[102] "BSgenome.Scerevisiae.UCSC.sacCer2"                 
[103] "BSgenome.Scerevisiae.UCSC.sacCer3"                 
[104] "BSgenome.Sscrofa.UCSC.susScr11"                    
[105] "BSgenome.Sscrofa.UCSC.susScr3"                     
[106] "BSgenome.Sscrofa.UCSC.susScr3.masked"              
[107] "BSgenome.Tgondii.ToxoDB.7.0"                       
[108] "BSgenome.Tguttata.UCSC.taeGut1"                    
[109] "BSgenome.Tguttata.UCSC.taeGut1.masked"             
[110] "BSgenome.Tguttata.UCSC.taeGut2"                    
[111] "BSgenome.Vvinifera.URGI.IGGP12Xv0"                 
[112] "BSgenome.Vvinifera.URGI.IGGP12Xv2"                 
[113] "BSgenome.Vvinifera.URGI.IGGP8X"                    </code></pre>
</div>
</div>
<p>We can use the <a href="http://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg38"><code>BSgenome.Hsapiens.UCSC.hg38</code></a> package to extract sequences from peaks.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-25_f30ca8c6edcadab50ef476ac744f8b48">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span><span class="op">)</span><span class="op">{</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">'BSgenome.Hsapiens.UCSC.hg38'</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The ENCODE project has thousands of datasets available for download. We’ll grab an arbitrary ATAC-seq dataset from here:</p>
<ul>
<li><a href="https://www.encodeproject.org/files/ENCFF994UIF/@@download/ENCFF994UIF.bed.gz" class="uri">https://www.encodeproject.org/files/ENCFF994UIF/@@download/ENCFF994UIF.bed.gz</a></li>
</ul>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-26_1be396288077ddbfa83c727c65aa2098">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">adrenal_atac</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span><span class="st">'https://www.encodeproject.org/files/ENCFF994UIF/@@download/ENCFF994UIF.bed.gz'</span>,</span>
<span>                                col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">adrenal_atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'chrom'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'score'</span>,<span class="st">'strand'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">adrenal_atac</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  chrom     start       end name        score strand    ``     ``     ``    ``
  &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
1 chr1  100037554 100039053 Peak_2668    1000 .      18.8  2883.  2879.    506
2 chr1  100037554 100039053 Peak_40256   1000 .       5.92  434.   431.   1016
3 chr1  100037554 100039053 Peak_57398   1000 .       4.71  278.   275.    811
4 chr1  100037554 100039053 Peak_78482   1000 .       3.77  174.   172.   1287
5 chr1  100046439 100046610 Peak_244701   146 .       2.75   14.7   12.9    83
6 chr1  100060446 100060954 Peak_138710   597 .       5.32   59.7   57.7   275</code></pre>
</div>
</div>
<p>We can convert the data to a <code>GRanges</code> object after ensureing that the column names are correct.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-27_98cdc0f33717ee10e0341d6908757308">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">adrenal_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="va">adrenal_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">adrenal_atac</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 6 metadata columns:
      seqnames              ranges strand |        name     score      &lt;NA&gt;
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;
  [1]     chr1 100037554-100039053      * |   Peak_2668      1000  18.84493
  [2]     chr1 100037554-100039053      * |  Peak_40256      1000   5.91615
  [3]     chr1 100037554-100039053      * |  Peak_57398      1000   4.70575
  [4]     chr1 100037554-100039053      * |  Peak_78482      1000   3.76879
  [5]     chr1 100046439-100046610      * | Peak_244701       146   2.74992
  [6]     chr1 100060446-100060954      * | Peak_138710       597   5.32355
           &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;
      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
  [1] 2882.5295 2878.5525       506
  [2]  433.6264  431.0128      1016
  [3]  277.7893  275.3472       811
  [4]  174.3991  172.1142      1287
  [5]   14.6544   12.9433        83
  [6]   59.7187   57.7244       275
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">adrenal_atac</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 273730</code></pre>
</div>
</div>
<p>We can extract the sequences from the peaks using the <a href="http://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg38"><code>BSgenome.Hsapiens.UCSC.hg38</code></a> package.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-28_7b6bcef284e14f2633783ba0c09304af">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">adrenal_atac_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span>, <span class="va">adrenal_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">adrenal_atac_seqs</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 6:
    width seq
[1]  1500 CAAAGGGTGACATCGGTCAGGTAAGGTATTTTTT...GATCGGAGGAGGACACTTTGAAGGGAAGAAAGG
[2]  1500 CAAAGGGTGACATCGGTCAGGTAAGGTATTTTTT...GATCGGAGGAGGACACTTTGAAGGGAAGAAAGG
[3]  1500 CAAAGGGTGACATCGGTCAGGTAAGGTATTTTTT...GATCGGAGGAGGACACTTTGAAGGGAAGAAAGG
[4]  1500 CAAAGGGTGACATCGGTCAGGTAAGGTATTTTTT...GATCGGAGGAGGACACTTTGAAGGGAAGAAAGG
[5]   172 TGGTCTTGAACTCCTGACCTCAGGTGATCCGCCC...GTTCCATATAAGTTTTAGGATTTAAAAGAAAAA
[6]   509 TGCAAGTTGAGCAGTCTGGAGCTGGTATGGCAGT...CCATGGTGTTATCTACAAAGGCCCAAAGGTTTA</code></pre>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">adrenal_atac_seqs</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 273730</code></pre>
</div>
</div>
<p>We can then write the sequences to a FASTA file.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-29_bd59124ebb9c6eb3175950e2aebf1a27">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">writeXStringSet</a></span><span class="op">(</span><span class="va">adrenal_atac_seqs</span>, <span class="st">'adrenal_atac_seqs.fa'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="common-peaks" class="level2" data-number="20.5"><h2 data-number="20.5" class="anchored" data-anchor-id="common-peaks">
<span class="header-section-number">20.5</span> Common peaks</h2>
<p>Let’s grab another dataset from the ENCODE project. This time, we’ll use a dataset from the <a href="https://www.encodeproject.org/">ENCODE Project</a>.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-30_7a06fe0e4db0079fc611af9c0f4b4880">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cardiac_atac</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span><span class="st">'https://www.encodeproject.org/files/ENCFF012BKZ/@@download/ENCFF012BKZ.bed.gz'</span>,</span>
<span>                                col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">cardiac_atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'chrom'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'name'</span>,<span class="st">'score'</span>,<span class="st">'strand'</span><span class="op">)</span></span>
<span><span class="va">cardiac_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="va">cardiac_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cardiac_atac</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 6 metadata columns:
      seqnames              ranges strand |        name     score      &lt;NA&gt;
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;
  [1]     chr1 100006549-100007015      * | Peak_126876       195   4.37594
  [2]     chr1 100006549-100007015      * | Peak_228935        62   2.54600
  [3]     chr1 100023516-100023841      * |  Peak_53548       741   9.38837
  [4]     chr1 100037565-100038984      * | Peak_140410       162   2.53385
  [5]     chr1 100037565-100038984      * | Peak_243001        56   1.78235
  [6]     chr1 100037565-100038984      * |   Peak_4079      1000  20.92354
           &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;
      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
  [1]  19.54719  17.50513       340
  [2]   6.23524   4.47481       104
  [3]  74.18277  71.72324       156
  [4]  16.24701  14.25361      1006
  [5]   5.63139   3.89871        82
  [6] 768.94092 765.21680       489
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Using what we know about overlaps, we can look for peaks that are common between the two datasets.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-31_f9ae680aa126bdd39a010a5f6f704e5e">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">adrenal_over_cardiac</span> <span class="op">&lt;-</span> <span class="va">adrenal_atac</span><span class="op">[</span><span class="va">adrenal_atac</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">cardiac_atac</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To export the GRanges as a BED file, we can use the <a href="http://bioconductor.org/packages/rtracklayer"><code>rtracklayer</code></a> package.</p>
<div class="cell" data-hash="atac_cache/html/unnamed-chunk-32_1d3de72755d8e35fd62e545269682df4">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="fu">export</span><span class="op">(</span><span class="va">adrenal_over_cardiac</span>, <span class="st">'adrenal_over_cardiac.bed'</span>, format<span class="op">=</span><span class="st">'bed'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Open the BED file in IGV. What do you see? What do you expect to see?</p>
<ul>
<li>Use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">GenomicRanges::intersect</a></code> method to find the common regions shared between the two datasets.</li>
<li>Use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">GenomicRanges::setdiff</a></code> method to find the regions that are unique to each dataset.</li>
<li>Use the <code><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html">GenomicRanges::disjoin</a></code> method to find all the disjoint regions between the two datasets.</li>
<li>What is the relationship between the number of disjoint regions and the number of peaks in the intersection and setdiff?</li>
</ul></section><section id="additional-work" class="level2" data-number="20.6"><h2 data-number="20.6" class="anchored" data-anchor-id="additional-work">
<span class="header-section-number">20.6</span> Additional work</h2>
<p>For those working extensively on ATAC-Seq, there is a great workflow/tutorial available from Thomas Carrol:</p>
<p>https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html</p>
<p>Feel free to work through it. In addition to the work above, there is also the <a href="http://bioconductor.org/packages/ATACseqQC"><code>ATACseqQC</code></a> package vignette that offers more than just QC. At least a couple more packages are available in <em>Bioconductor</em>.</p>
</section><section id="appendix" class="level2 unnumbered page-columns page-full"><h2 class="unnumbered anchored" data-anchor-id="appendix">Appendix</h2>
<section id="session-info" class="level3 unnumbered"><h3 class="unnumbered anchored" data-anchor-id="session-info">Session info</h3>
<div class="cell" data-hash="atac_cache/html/sessionInfo_33d5ccc41c27ca559bb5bf668b19499d">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.0 (2023-04-21)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] rtracklayer_1.60.0                     
 [2] heatmaps_1.24.0                        
 [3] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
 [4] GenomicFeatures_1.52.1                 
 [5] AnnotationDbi_1.62.2                   
 [6] dplyr_1.1.2                            
 [7] ggplot2_3.4.2                          
 [8] GenomicAlignments_1.36.0               
 [9] Rsamtools_2.16.0                       
[10] Biostrings_2.68.1                      
[11] XVector_0.40.0                         
[12] SummarizedExperiment_1.30.2            
[13] Biobase_2.60.0                         
[14] MatrixGenerics_1.12.2                  
[15] matrixStats_1.0.0                      
[16] GenomicRanges_1.52.0                   
[17] GenomeInfoDb_1.36.1                    
[18] IRanges_2.34.1                         
[19] S4Vectors_0.38.1                       
[20] BiocGenerics_0.46.0                    
[21] knitr_1.43                             

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0        EBImage_4.42.0          blob_1.2.4             
 [4] filelock_1.0.2          bitops_1.0-7            fastmap_1.1.1          
 [7] RCurl_1.98-1.12         BiocFileCache_2.8.0     XML_3.99-0.14          
[10] digest_0.6.31           lifecycle_1.0.3         KEGGREST_1.40.0        
[13] RSQLite_2.3.1           magrittr_2.0.3          compiler_4.3.0         
[16] rlang_1.1.1             progress_1.2.2          tools_4.3.0            
[19] plotrix_3.8-2           utf8_1.2.3              yaml_2.3.7             
[22] prettyunits_1.1.1       S4Arrays_1.0.4          htmlwidgets_1.6.2      
[25] bit_4.0.5               curl_5.0.1              DelayedArray_0.26.6    
[28] RColorBrewer_1.1-3      xml2_1.3.5              KernSmooth_2.23-21     
[31] abind_1.4-5             BiocParallel_1.34.2     withr_2.5.0            
[34] grid_4.3.0              fansi_1.0.4             colorspace_2.1-0       
[37] scales_1.2.1            biomaRt_2.56.1          cli_3.6.1              
[40] rmarkdown_2.23          crayon_1.5.2            generics_0.1.3         
[43] rstudioapi_0.14         rjson_0.2.21            httr_1.4.6             
[46] DBI_1.1.3               cachem_1.0.8            stringr_1.5.0          
[49] zlibbioc_1.46.0         parallel_4.3.0          tiff_0.1-11            
[52] restfulr_0.0.15         vctrs_0.6.3             Matrix_1.5-4.1         
[55] jsonlite_1.8.7          fftwtools_0.9-11        hms_1.1.3              
[58] bit64_4.0.5             jpeg_0.1-10             locfit_1.5-9.8         
[61] glue_1.6.2              codetools_0.2-19        stringi_1.7.12         
[64] gtable_0.3.3            BiocIO_1.10.0           munsell_0.5.0          
[67] tibble_3.2.1            pillar_1.9.0            rappdirs_0.3.3         
[70] htmltools_0.5.5         GenomeInfoDbData_1.2.10 R6_2.5.1               
[73] dbplyr_2.3.2            evaluate_0.21           lattice_0.21-8         
[76] png_0.1-8               memoise_2.0.1           xfun_0.39              
[79] pkgconfig_2.0.3        </code></pre>
</div>
</div>
</section><section id="macs2" class="level3 unnumbered page-columns page-full"><h3 class="unnumbered anchored" data-anchor-id="macs2">MACS2</h3>
<div class="page-columns page-full"><p>The MACS2 package is a commonly-used package for calling peaks. Installation and other details are available<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p><div class="no-row-height column-margin column-container"><li id="fn1"><p><sup>1</sup>&nbsp;https://github.com/taoliu/MACS</p></li></div></div>
<pre><code>pip install macs2</code></pre>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Buenrostro2013-dz" class="csl-entry" role="listitem">
Buenrostro, Jason D, Paul G Giresi, Lisa C Zaba, Howard Y Chang, and William J Greenleaf. 2013. <span>“Transposition of Native Chromatin for Fast and Sensitive Epigenomic Profiling of Open Chromatin, <span class="nocase">DNA-binding</span> Proteins and Nucleosome Position.”</span> <em>Nature Methods</em> 10 (12): 1213–18. <a href="https://doi.org/10.1038/nmeth.2688">https://doi.org/10.1038/nmeth.2688</a>.
</div>
<div id="ref-Buenrostro2015-dz" class="csl-entry" role="listitem">
Buenrostro, Jason D, Beijing Wu, Howard Y Chang, and William J Greenleaf. 2015. <span>“<span class="nocase">ATAC-seq</span>: A Method for Assaying Chromatin Accessibility <span>Genome-Wide</span>.”</span> <em>Current Protocols in Molecular Biology / Edited by Frederick M. Ausubel ... [Et Al.]</em> 109 (January): 21.29.1–9. <a href="https://doi.org/10.1002/0471142727.mb2129s109">https://doi.org/10.1002/0471142727.mb2129s109</a>.
</div>
<div id="ref-Crawford2006-nt" class="csl-entry" role="listitem">
Crawford, Gregory E, Sean Davis, Peter C Scacheri, Gabriel Renaud, Mohamad J Halawi, Michael R Erdos, Roland Green, Paul S Meltzer, Tyra G Wolfsberg, and Francis S Collins. 2006. <span>“<span class="nocase">DNase-chip</span>: A High-Resolution Method to Identify <span>DNase</span> <span>I</span> Hypersensitive Sites Using Tiled Microarrays.”</span> <em>Nature Methods</em> 3 (7): 503–9. <a href="http://www.ncbi.nlm.nih.gov/pubmed/16791207?dopt=AbstractPlus">http://www.ncbi.nlm.nih.gov/pubmed/16791207?dopt=AbstractPlus</a>.
</div>
<div id="ref-Crawford2006-bk" class="csl-entry" role="listitem">
Crawford, Gregory E, Ingeborg E Holt, James Whittle, Bryn D Webb, Denise Tai, Sean Davis, Elliott H Margulies, et al. 2006. <span>“Genome-Wide Mapping of <span>DNase</span> Hypersensitive Sites Using Massively Parallel Signature Sequencing (<span>MPSS</span>).”</span> <em>Genome Research</em> 16 (1): 123–31. <a href="http://www.ncbi.nlm.nih.gov/pubmed/16344561?dopt=AbstractPlus">http://www.ncbi.nlm.nih.gov/pubmed/16344561?dopt=AbstractPlus</a>.
</div>
<div id="ref-Lawrence2013-he" class="csl-entry" role="listitem">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em>PLoS Computational Biology</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-morgan2016rsamtools" class="csl-entry" role="listitem">
Morgan, Martin, Herve Pages, V Obenchain, and N Hayden. 2016. <span>“Rsamtools: Binary Alignment (BAM), FASTA, Variant Call (BCF), and Tabix File Import.”</span> <em>R Package Version</em> 1 (0): 677–89.
</div>
<div id="ref-Tsompana2014-yh" class="csl-entry" role="listitem">
Tsompana, Maria, and Michael J Buck. 2014. <span>“Chromatin Accessibility: A Window into the Genome.”</span> <em>Epigenetics &amp; Chromatin</em> 7 (1): 33. <a href="https://doi.org/10.1186/1756-8935-7-33">https://doi.org/10.1186/1756-8935-7-33</a>.
</div>
</div>
</section></section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./ranges_and_signals.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Genomic ranges and features</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="./license.html">License</a>
  </li>  
</ul>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>