<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Sean Davis">
<meta name="dcterms.date" content="2024-06-01">
<meta name="description" content="A collection of chapters more than a book, but….">
<title>28&nbsp; Single Cell RNA-seq Analysis – The RBioc Book</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../references.html" rel="next">
<link href="../atac-seq/atac-seq.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-180dc90b24616f542b21b8b54fc3f545.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-5a15bea12966ac2249ff2e96cb8142e0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-KLLV1GCF4E"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-KLLV1GCF4E', { 'anonymize_ip': true});
</script><style>html{ scroll-behavior: smooth; }</style>
<script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">The RBioc Book</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../intro.html"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../data_structures_overview.html"> 
<span class="menu-text">Data Structures</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/seandavi/RBiocBook" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="../The-RBioc-Book.pdf">
              <i class="bi bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="../The-RBioc-Book.epub">
              <i class="bi bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
</div>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-1" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-1">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../geoquery.html">Bioconductor</a></li><li class="breadcrumb-item"><a href="../single_cell/setup.html"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Single Cell RNA-seq Analysis</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../intro_to_rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../r_intro_mechanics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R mechanics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Up and Running with R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../packages_and_dice.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Packages and more dice</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../reading_and_writing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Reading and writing data files</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../data_structures_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R Data Structures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../vectors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Vectors</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../matrices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lists.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lists</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../dataframes_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Data Frames</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../factors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Factors</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../eda_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exploratory data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../dplyr_intro_msleep.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to dplyr: mammal sleep dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../eda_and_univariate_brfss.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Behavioral Risk Factor Surveillance System</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../visualization_guide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Self-Guided Data Visualization in R</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">statististics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../norm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Working with distribution functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../t-stats-and-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">The t-statistic and t-distribution</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../kmeans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">K-means clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Machine Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../machine_learning/intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../machine_learning/models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Supervised Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../machine_learning/mlr3verse_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">The mlr3verse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../machine_learning/machine_learning_mlr3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Examples</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Bioconductor</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../geoquery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Accessing and working with public omics data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../bioc-summarizedexperiment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Introduction to <code>SummarizedExperiment</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../genomic_ranges_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Genomic Ranges Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../ranges_and_signals.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Genomic ranges and features</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../ranges_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Ranges Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../atac-seq/atac-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Genomic ranges and ATAC-Seq</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/setup.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Single Cell RNA-seq Analysis</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../appendix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Appendix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../git_and_github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Git and GitHub</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../additional_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Additional resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../dataviz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../matrix_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">E</span>&nbsp; <span class="chapter-title">Matrix Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">28.1</span> Introduction</a></li>
  <li>
<a href="#setup-and-data-loading" id="toc-setup-and-data-loading" class="nav-link" data-scroll-target="#setup-and-data-loading"><span class="header-section-number">28.2</span> Setup and Data Loading</a>
  <ul class="collapse">
<li><a href="#loading-essential-libraries" id="toc-loading-essential-libraries" class="nav-link" data-scroll-target="#loading-essential-libraries"><span class="header-section-number">28.2.1</span> Loading Essential Libraries</a></li>
  <li><a href="#exploring-available-datasets" id="toc-exploring-available-datasets" class="nav-link" data-scroll-target="#exploring-available-datasets"><span class="header-section-number">28.2.2</span> Exploring Available Datasets</a></li>
  </ul>
</li>
  <li>
<a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction"><span class="header-section-number">28.3</span> Dimensionality reduction</a>
  <ul class="collapse">
<li><a href="#feature-selection-identifying-highly-variable-genes" id="toc-feature-selection-identifying-highly-variable-genes" class="nav-link" data-scroll-target="#feature-selection-identifying-highly-variable-genes"><span class="header-section-number">28.3.1</span> Feature Selection: Identifying Highly Variable Genes</a></li>
  <li><a href="#non-linear-dimensionality-reduction-with-t-sne" id="toc-non-linear-dimensionality-reduction-with-t-sne" class="nav-link" data-scroll-target="#non-linear-dimensionality-reduction-with-t-sne"><span class="header-section-number">28.3.2</span> Non-linear Dimensionality Reduction with t-SNE</a></li>
  </ul>
</li>
  <li><a href="#why-t-sne-after-pca" id="toc-why-t-sne-after-pca" class="nav-link" data-scroll-target="#why-t-sne-after-pca"><span class="header-section-number">28.4</span> Why t-SNE after PCA</a></li>
  <li><a href="#visualizing-cell-types-and-clusters" id="toc-visualizing-cell-types-and-clusters" class="nav-link" data-scroll-target="#visualizing-cell-types-and-clusters"><span class="header-section-number">28.4.1</span> Visualizing Cell Types and Clusters</a></li>
  <li><a href="#finding-marker-genes" id="toc-finding-marker-genes" class="nav-link" data-scroll-target="#finding-marker-genes"><span class="header-section-number">28.5</span> Finding Marker Genes</a></li>
  <li><a href="#interactive-visualization-with-isee" id="toc-interactive-visualization-with-isee" class="nav-link" data-scroll-target="#interactive-visualization-with-isee"><span class="header-section-number">28.6</span> Interactive Visualization with iSEE</a></li>
  <li>
<a href="#bonus-working-with-10x-genomics-data" id="toc-bonus-working-with-10x-genomics-data" class="nav-link" data-scroll-target="#bonus-working-with-10x-genomics-data"><span class="header-section-number">28.7</span> Bonus: Working with 10x Genomics Data</a>
  <ul class="collapse">
<li><a href="#understanding-10x-genomics-output-formats" id="toc-understanding-10x-genomics-output-formats" class="nav-link" data-scroll-target="#understanding-10x-genomics-output-formats"><span class="header-section-number">28.7.1</span> Understanding 10x Genomics Output Formats</a></li>
  <li><a href="#loading-10x-data-from-matrix-market-files" id="toc-loading-10x-data-from-matrix-market-files" class="nav-link" data-scroll-target="#loading-10x-data-from-matrix-market-files"><span class="header-section-number">28.7.2</span> Loading 10x Data from Matrix Market Files</a></li>
  <li><a href="#loading-from-hdf5-files" id="toc-loading-from-hdf5-files" class="nav-link" data-scroll-target="#loading-from-hdf5-files"><span class="header-section-number">28.7.3</span> Loading from HDF5 Files</a></li>
  <li><a href="#essential-quality-control-after-loading-raw-data" id="toc-essential-quality-control-after-loading-raw-data" class="nav-link" data-scroll-target="#essential-quality-control-after-loading-raw-data"><span class="header-section-number">28.7.4</span> Essential Quality Control After Loading Raw Data</a></li>
  <li><a href="#filtering-low-quality-cells-and-genes" id="toc-filtering-low-quality-cells-and-genes" class="nav-link" data-scroll-target="#filtering-low-quality-cells-and-genes"><span class="header-section-number">28.7.5</span> Filtering Low-Quality Cells and Genes</a></li>
  </ul>
</li>
  <li><a href="#further-reading-and-resources" id="toc-further-reading-and-resources" class="nav-link" data-scroll-target="#further-reading-and-resources"><span class="header-section-number">28.8</span> Further Reading and Resources</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/seandavi/RBiocBook/edit/main/single_cell/setup.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/seandavi/RBiocBook/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/seandavi/RBiocBook/blob/main/single_cell/setup.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../geoquery.html">Bioconductor</a></li><li class="breadcrumb-item"><a href="../single_cell/setup.html"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Single Cell RNA-seq Analysis</span></a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Single Cell RNA-seq Analysis</span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source" data-quarto-source-url="https://github.com/seandavi/RBiocBook/blob/main/single_cell/setup.qmd"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://seandavi.github.io/">Sean Davis</a> <a href="mailto:seandavi@gmail.com" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://medschool.cuanschutz.edu/">
            University of Colorado<br>Anschutz School of Medicine
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">June 1, 2024</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">June 18, 2025</p>
    </div>
  </div>
    
  </div>
  


</header><section id="introduction" class="level2" data-number="28.1"><h2 data-number="28.1" class="anchored" data-anchor-id="introduction">
<span class="header-section-number">28.1</span> Introduction</h2>
<p>Single-cell RNA sequencing (scRNA-seq) is a powerful technique that allows us to measure gene expression in individual cells rather than bulk tissue samples. This capability has revolutionized our understanding of cellular heterogeneity, developmental processes, and disease mechanisms. Unlike traditional bulk RNA-seq, which provides an average expression profile across all cells in a sample, scRNA-seq reveals the unique molecular signatures of individual cells, enabling identification of distinct cell types, states, and trajectories.</p>
<p>This tutorial will walk through a typical scRNA-seq analysis workflow using R and the Bioconductor ecosystem. We’ll cover data loading, quality control, normalization, dimensionality reduction, clustering, and visualization - the fundamental steps that transform raw sequencing data into biological insights.</p>
</section><section id="setup-and-data-loading" class="level2 page-columns page-full" data-number="28.2"><h2 data-number="28.2" class="anchored" data-anchor-id="setup-and-data-loading">
<span class="header-section-number">28.2</span> Setup and Data Loading</h2>
<section id="loading-essential-libraries" class="level3" data-number="28.2.1"><h3 data-number="28.2.1" class="anchored" data-anchor-id="loading-essential-libraries">
<span class="header-section-number">28.2.1</span> Loading Essential Libraries</h3>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load essential libraries for single-cell RNA-seq analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We start by loading three essential R packages for single-cell analysis:</p>
<ul>
<li>
<strong>scRNAseq</strong>: Provides access to curated single-cell RNA-seq datasets, making it easy to practice analysis techniques on real data. At the time of writing, it includes more than 50 datasets covering various tissues, organisms, and experimental conditions. See <span class="citation" data-cites="risso2024scRNAseq">Risso and Cole (<a href="../references.html#ref-risso2024scRNAseq" role="doc-biblioref">2024</a>)</span>
</li>
<li>
<strong>scran</strong>: Contains methods for normalization, feature selection, and clustering specifically designed for single-cell data. See <span class="citation" data-cites="lun2016">Lun, McCarthy, and Marioni (<a href="../references.html#ref-lun2016" role="doc-biblioref">2016</a>)</span>.</li>
<li>
<strong>scater</strong>: Offers tools for quality control, visualization, and exploratory data analysis. See <span class="citation" data-cites="McCarthy2017scater">McCarthy et al. (<a href="../references.html#ref-McCarthy2017scater" role="doc-biblioref">2017</a>)</span>.</li>
</ul>
<p>Single-cell data has unique characteristics (sparsity, high dimensionality, technical noise) that require specialized tools. These packages form the backbone of the Bioconductor single-cell analysis ecosystem and provide methods that account for the specific challenges of single-cell data.</p>
</section><section id="exploring-available-datasets" class="level3 page-columns page-full" data-number="28.2.2"><h3 data-number="28.2.2" class="anchored" data-anchor-id="exploring-available-datasets">
<span class="header-section-number">28.2.2</span> Exploring Available Datasets</h3>
<p>We are going to use the <code>scRNAseq</code> package to access a curated collection of single-cell datasets. This package provides a convenient way to load and explore publicly available datasets that have been processed and annotated.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/surveyDatasets.html">surveyDatasets</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scRNAseq/man/surveyDatasets.html">surveyDatasets()</a></code> function queries the scRNAseq package database to show all available datasets. This creates a simple table of publicly available single-cell datasets that have been processed and made easily accessible.</p>
<p>Having access to well-curated datasets is useful for learning and benchmarking. These datasets come from published studies and represent different tissues, conditions, and experimental designs, allowing us to practice on real biological data without having to process raw sequencing files.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 10 rows and 15 columns
                     name     version        path                 object
              &lt;character&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
1       aztekin-tail-2019  2023-12-14          NA single_cell_experiment
2  splicing-demonstrati..  2023-12-20          NA single_cell_experiment
3      marques-brain-2016  2023-12-19          NA single_cell_experiment
4   grun-bone_marrow-2016  2023-12-14          NA single_cell_experiment
5         giladi-hsc-2018  2023-12-21      crispr single_cell_experiment
6         giladi-hsc-2018  2023-12-21         rna single_cell_experiment
7     macosko-retina-2015  2023-12-19          NA single_cell_experiment
8        messmer-esc-2019  2023-12-19          NA single_cell_experiment
9  ernst-spermatogenesi..  2023-12-21  cellranger single_cell_experiment
10 ernst-spermatogenesi..  2023-12-21  emptydrops single_cell_experiment
                    title            description taxonomy_id   genome      rows
              &lt;character&gt;            &lt;character&gt;      &lt;List&gt;   &lt;List&gt; &lt;integer&gt;
1  Identification of a .. Identification of a ..        8355 Xenla9.1     31535
2  [reprocessed, subset.. [reprocessed, subset..       10090   GRCm38     54448
3  Oligodendrocyte hete.. Oligodendrocyte hete..       10090   GRCm38     23556
4  De Novo Prediction o.. De Novo Prediction o..       10090   GRCm38     23536
5  Single-cell characte.. Single-cell characte..       10090  MGSCv37        30
6  Single-cell characte.. Single-cell characte..       10090  MGSCv37     27389
7  Highly Parallel Geno.. Highly Parallel Geno..       10090   GRCm38     24658
8  Transcriptional Hete.. Transcriptional Hete..        9606   GRCh38     58302
9  Staged developmental.. Staged developmental..       10090   GRCm38     33226
10 Staged developmental.. Staged developmental..       10090   GRCm38     32105
     columns            assays                               column_annotations
   &lt;integer&gt;            &lt;List&gt;                                           &lt;List&gt;
1      13199            counts sample,DevelopmentalStage,DaysPostAmputation,...
2       2325 spliced,unspliced                                         celltype
3       5069            counts           source_name,age,inferred cell type,...
4       1915            counts                                  sample,protocol
5      24070            counts                              amplification.batch
6      81408            counts               Amp_batch_ID,Seq_batch_ID,tier,...
7      49300            counts                                          cluster
8       1344            counts    Source Name,phenotype,single cell quality,...
9      53510            counts                       Sample,Barcode,Library,...
10     68937            counts                       Sample,Barcode,Library,...
   reduced_dimensions alternative_experiments
               &lt;List&gt;                  &lt;List&gt;
1                UMAP                        
2                                            
3                                            
4                                            
5                                            
6                                            
7                                            
8                                        ERCC
9                                            
10                                           
                                                                    sources
                                                       &lt;SplitDataFrameList&gt;
1                            ArrayExpress:E-MTAB-7716:NA,PubMed:31097661:NA
2                      GEO:GSE109033:NA,GEO:GSM2928341:NA,SRA:SRR6459157:NA
3                                        GEO:GSE75330:NA,PubMed:27284195:NA
4                                        GEO:GSE76983:NA,PubMed:27345837:NA
5                                        PubMed:29915358:NA,GEO:GSE11349:NA
6                                        PubMed:29915358:NA,GEO:GSE92575:NA
7  GEO:GSE63472:NA,PubMed:26000488:NA,URL:http://mccarrolllab...:2024-02-23
8                            PubMed:30673604:NA,ArrayExpress:E-MTAB-6819:NA
9                            ArrayExpress:E-MTAB-6946:NA,PubMed:30890697:NA
10                           ArrayExpress:E-MTAB-6946:NA,PubMed:30890697:NA</code></pre>
</div>
</div>
<p>Each row represents a different study. The ‘columns’ field shows the number of cells, while ‘rows’ shows the number of genes. This helps you choose datasets appropriate for your computational resources and analysis goals. This little lightweight database is not the richest way to explore datasets, but it is a good starting point. In this tutorial, we will use the <code>ZeiselBrainData</code> dataset, which contains single-cell RNA-seq data from mouse brain tissue (<span class="citation" data-cites="zeiselCellTypesMouse2015">Zeisel et al. (<a href="../references.html#ref-zeiselCellTypesMouse2015" role="doc-biblioref">2015</a>)</span>).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Abstract: Zeisel Brain Dataset
</div>
</div>
<div class="callout-body-container callout-body">
<p>The mammalian cerebral cortex supports cognitive functions such as sensorimotor integration, memory, and social behaviors. Normal brain function relies on a diverse set of differentiated cell types, including neurons, glia, and vasculature. Here, we have used large-scale single-cell RNA sequencing (RNA-seq) to classify cells in the mouse somatosensory cortex and hippocampal CA1 region. We found 47 molecularly distinct subclasses, comprising all known major cell types in the cortex. We identified numerous marker genes, which allowed alignment with known cell types, morphology, and location. We found a layer I interneuron expressing Pax6 and a distinct postmitotic oligodendrocyte subclass marked by Itpr2. Across the diversity of cortical cell types, transcription factors formed a complex, layered regulatory code, suggesting a mechanism for the maintenance of adult cell type identity.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What is the <code>sce</code> object? It is a <code>SingleCellExperiment</code> object, which is a data structure specifically designed for single-cell RNA-seq data. As with other Bioconductor data structures, it is built on top of the <code>SummarizedExperiment</code> class (see <a href="#fig-scRNAseq-sce" class="quarto-xref">Figure&nbsp;<span>28.1</span></a>), which allows us to store not only the count matrix but also metadata about cells and features (genes).</p>
<div id="fig-scRNAseq-sce" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-scRNAseq-sce-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-019-0654-x/MediaObjects/41592_2019_654_Fig2_HTML.png" class="img-fluid quarto-figure quarto-figure-center figure-img">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-scRNAseq-sce-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;28.1: The SingleCellExperiment object.
</figcaption></figure>
</div>
<p>Printing the <code>sce</code> object gives us a summary of the dataset:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 20006 3005 
metadata(0):
assays(1): counts
rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l
rowData names(1): featureType
colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12
  1772058148_F03
colData names(9): tissue group # ... level1class level2class
reducedDimNames(0):
mainExpName: gene
altExpNames(2): repeat ERCC</code></pre>
</div>
</div>
<p>The <code>SingleCellExperiment</code> object contains:</p>
<ul>
<li>
<strong>Assays</strong>: The main data matrix (counts) and any additional matrices (e.g., normalized counts).</li>
<li>
<strong>Row metadata</strong>: Information about genes (e.g., gene names, IDs).</li>
<li>
<strong>Column metadata</strong>: Information about cells (e.g., cell barcodes, tissue type, experimental conditions).</li>
</ul>
<p>It <strong>can</strong> also contain additional information such as dimensionality reductions (e.g., PCA, t-SNE), clustering results, and quality control metrics.</p>
<p>In fact, we are going to add some of these additional information in the next steps. First, we add a new assay with the log-normalized counts, which is a common preprocessing step in single-cell RNA-seq analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts()</a></code> function performs log-normalization of the count data. Raw count data from scRNA-seq experiments contains technical variation due to differences in sequencing depth between cells. Some cells might have been sequenced more deeply than others, leading to higher total counts that don’t reflect true biological differences. Log-normalization addresses this by:</p>
<ol type="1">
<li>
<strong>Size factor normalization</strong>: Scaling each cell’s counts by a size factor that accounts for differences in sequencing dept. There are several methods to calculate size factors, but the most common is the median of ratios method, which divides each cell’s counts by the median count for each gene across all cells. This helps to adjust for differences in library size.</li>
<li>
<strong>Log transformation</strong>: Taking log(count + 1) to stabilize variance and make the data more normally distributed</li>
</ol>
<p>For each gene g in cell i, the log-normalized expression becomes:</p>
<p><span class="math display">\[log_2(count_{g_i} × {size factor}_{i} + 1)\]</span></p>
<p>This log transformation is useful because the next steps in the downstream analyses (like clustering and dimensionality reduction) assume that the data is approximately normally distributed (or at least bell-shaped). Log-normalization helps achieve this by reducing the impact of outliers and making the data more suitable for PCA and visualization.</p>
</section></section><section id="dimensionality-reduction" class="level2 page-columns page-full" data-number="28.3"><h2 data-number="28.3" class="anchored" data-anchor-id="dimensionality-reduction">
<span class="header-section-number">28.3</span> Dimensionality reduction</h2>
<p>Single-cell datasets are extremely high-dimensional - typically containing 15,000-30,000 genes per cell. This creates several challenges: computational complexity, noise dominance, and visualization difficulties. Dimensionality reduction techniques help us identify the most informative patterns in the data while reducing noise and computational burden.</p>
<section id="feature-selection-identifying-highly-variable-genes" class="level3 page-columns page-full" data-number="28.3.1"><h3 data-number="28.3.1" class="anchored" data-anchor-id="feature-selection-identifying-highly-variable-genes">
<span class="header-section-number">28.3.1</span> Feature Selection: Identifying Highly Variable Genes</h3>
<p>Before applying dimensionality reduction, we need to identify the most informative genes. Highly variable genes (HVGs) are those that show significant variation across cells, indicating they may be biologically relevant. These genes are often more informative for downstream analyses like clustering and visualization.</p>
<p>Not all genes are equally informative for distinguishing cell types. Genes that show little variation across cells (like housekeeping genes) don’t help us identify distinct cell populations. Highly variable genes are more likely to:</p>
<ul>
<li>Reflect biological differences between cell types</li>
<li>Capture developmental or activation states</li>
<li>Represent responses to environmental conditions</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/">scran</a></span><span class="op">)</span></span>
<span><span class="va">top_var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span><span class="op">(</span><span class="va">sce</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs()</a></code> function identifies the top 2000 highly variable genes based on their variance across cells. Choosing a subset of highly variable genes reduces the dimensionality of the dataset while retaining the most informative features for downstream analyses.</p>
<p>The algorithm models the relationship between gene expression mean and variance, then identifies genes that show more variation than expected based on their expression level. This accounts for the fact that highly expressed genes naturally show more variance.</p>
<div class="cell page-columns page-full">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualizing the highly variable genes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/">scran</a></span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualizing the fit:</span></span>
<span><span class="va">fit.mv</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">dec</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">fit.mv</span><span class="op">$</span><span class="va">mean</span>, <span class="va">fit.mv</span><span class="op">$</span><span class="va">var</span>, xlab<span class="op">=</span><span class="st">"Mean of log-expression"</span>,</span>
<span>    ylab<span class="op">=</span><span class="st">"Variance of log-expression"</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span><span class="op">(</span><span class="va">fit.mv</span><span class="op">$</span><span class="fu">trend</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"dodgerblue"</span>, add<span class="op">=</span><span class="cn">TRUE</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span></span>
<span>    <span class="va">fit.mv</span><span class="op">$</span><span class="va">mean</span><span class="op">[</span><span class="va">top_var_genes</span><span class="op">]</span>, <span class="va">fit.mv</span><span class="op">$</span><span class="va">var</span><span class="op">[</span><span class="va">top_var_genes</span><span class="op">]</span>, </span>
<span>    col<span class="op">=</span><span class="st">"red"</span>, pch<span class="op">=</span><span class="fl">16</span>, cex<span class="op">=</span><span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-scRNAseq-hvg" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-scRNAseq-hvg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="setup_files/figure-html/fig-scRNAseq-hvg-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-scRNAseq-hvg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;28.2: A plot of the mean vs variance of log-expression for all the genes in the dataset, with the highly variable genes highlighted in red. Here, the x-axis represents the mean log-expression of each gene, and the y-axis represents the variance of log-expression. The blue curve represents the trend of variance as a function of mean expression, while the red points represent the highly variable genes.
</figcaption></figure>
</div>
</div>
</div>
<p><a href="#fig-scRNAseq-hvg" class="quarto-xref">Figure&nbsp;<span>28.2</span></a> shows the mean vs.&nbsp;variance of log-expression for all genes in the dataset, with highly variable genes highlighted in red. The blue curve represents the trend of variance as a function of mean expression, while the red points represent the highly variable genes.</p>
<p>We next proceed to calculating PCA using the subset of highly variable genes. PCA is a linear dimensionality reduction technique that finds the directions of maximum variance in the data. It’s computationally efficient and provides a good foundation for further analysis. The first few principal components capture the most important patterns of variation in the dataset.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why PCA first?
</div>
</div>
<div class="callout-body-container callout-body">
<p>PCA is a widely <em>linear</em> dimensionality reduction technique that:</p>
<ul>
<li>Reduces dimensionality while preserving variance</li>
<li>Helps visualize high-dimensional data in 2D or 3D</li>
<li>Serves as a preprocessing step for other methods like t-SNE or UMAP</li>
<li>Is computationally efficient and easy to interpret</li>
</ul>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Perform PCA on the highly variable genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span> <span class="co"># See below.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html">fixedPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset.row<span class="op">=</span><span class="va">top_var_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDimNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PCA"</code></pre>
</div>
</div>
<p>Note that we assign the output of <code><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html">fixedPCA()</a></code> to the <code>sce</code> object. This function performs PCA on the highly variable genes we identified earlier, reducing the dimensionality of the dataset while retaining the most informative features. Because dimensionality reduction is a common step in single-cell RNA-seq analysis, we can store the PCA results in the <code>reducedDims</code> slot of the <code>SingleCellExperiment</code> object. This allows us to easily access and visualize the PCA coordinates later.</p>
</section><section id="non-linear-dimensionality-reduction-with-t-sne" class="level3" data-number="28.3.2"><h3 data-number="28.3.2" class="anchored" data-anchor-id="non-linear-dimensionality-reduction-with-t-sne">
<span class="header-section-number">28.3.2</span> Non-linear Dimensionality Reduction with t-SNE</h3>
<p>After PCA, we can apply non-linear dimensionality reduction techniques like t-SNE (t-distributed Stochastic Neighbor Embedding) to visualize the data in 2D or 3D. t-SNE is particularly useful for revealing local structure and separating distinct cell populations.</p>
</section></section><section id="why-t-sne-after-pca" class="level2 {.callout-tip)" data-number="28.4"><h2 data-number="28.4" class="anchored" data-anchor-id="why-t-sne-after-pca">
<span class="header-section-number">28.4</span> Why t-SNE after PCA</h2>
<p>This is a two-step approach: 1. <strong>PCA first</strong>: Reduces noise and computational burden while preserving global structure 2. <strong>t-SNE second</strong>: Reveals local neighborhood structure and separates distinct cell populations</p>
</section><div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE()</a></code> function applies t-SNE to the PCA results, creating a 2D representation of the data. The <code>dimred="PCA"</code> argument specifies that we want to use the PCA coordinates as input for t-SNE.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nn.clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"TSNE"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">nn.clusters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>nn.clusters
  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 
152 126 190 286 131 191  63  87 113 328  88  97 155  45 138  40  71  42  61  92 
 21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36 
 54  64  35  62  58  32  27  32  20  16  12  20  28  15  18  16 </code></pre>
</div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scran/man/clusterCells.html">clusterCells()</a></code> function uses graph-based clustering to identify groups of similar cells. Graph-based clustering was <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#cluster-the-cells">popularized by the Seurat package</a>. It builds a nearest-neighbor graph and then finds communities within this graph. The resulting clusters often correspond to distinct cell types or states. The <code>use.dimred = "TSNE"</code> argument specifies that we want to use the t-SNE coordinates for clustering. This is an important choide because t-SNE captures local structure and can reveal distinct cell populations that may not be apparent in the original high-dimensional space. There are other clustering methods available, such as k-means or hierarchical clustering, but graph-based clustering is particularly effective for single-cell data due to its ability to handle complex, non-linear relationships.</p>
<section id="visualizing-cell-types-and-clusters" class="level3 page-columns page-full" data-number="28.4.1"><h3 data-number="28.4.1" class="anchored" data-anchor-id="visualizing-cell-types-and-clusters">
<span class="header-section-number">28.4.1</span> Visualizing Cell Types and Clusters</h3>
<p>In this particular dataset, we have pre-defined cell type annotations that we can use to color the t-SNE plot. The <code>level1class</code> and <code>level2class</code> columns in the metadata contain these annotations. We can use these annotations to color the t-SNE plot and visualize how well the clustering aligns with known cell types.</p>
<p>We can cluster the cells based on their t-SNE coordinates and visualize the results. The <code><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim()</a></code> function from the <code>scater</code> package allows us to create a scatter plot of the t-SNE coordinates, coloring the points by their cluster assignments or cell type annotations.</p>
<div class="cell page-columns page-full">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nn.clusters</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-level1-class" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" width="672">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-level1-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="setup_files/figure-html/fig-level1-class-1.png" id="fig-level1-class" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig quarto-uncaptioned margin-caption" id="fig-level1-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;28.3
</figcaption></figure>
</div>
</div>
</div>
<p>We can also visualize the second level of cell type annotations.</p>
<div class="cell page-columns page-full">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span>, colour_by<span class="op">=</span><span class="st">"level2class"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-level2-class" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" width="672">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-level2-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="setup_files/figure-html/fig-level2-class-1.png" id="fig-level2-class" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig quarto-uncaptioned margin-caption" id="fig-level2-class-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;28.4
</figcaption></figure>
</div>
</div>
</div>
<p>And add additional metadata such as tissue type to the plot. The <code>shape_by</code> argument allows us to differentiate points by another categorical variable, such as tissue type.</p>
<div class="cell page-columns page-full">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"TSNE"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span>, shape_by<span class="op">=</span><span class="st">"tissue"</span>, <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-level1-tissue" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" width="672">
<figure class="quarto-float quarto-float-fig figure page-columns page-full"><div aria-describedby="fig-level1-tissue-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="setup_files/figure-html/fig-level1-tissue-1.png" id="fig-level1-tissue" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig quarto-uncaptioned margin-caption" id="fig-level1-tissue-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;28.5
</figcaption></figure>
</div>
</div>
</div>
</section><section id="finding-marker-genes" class="level2" data-number="28.5"><h2 data-number="28.5" class="anchored" data-anchor-id="finding-marker-genes">
<span class="header-section-number">28.5</span> Finding Marker Genes</h2>
<p>Marker genes are genes that are differentially expressed in a specific cell type or cluster compared to others. Identifying marker genes helps us understand the molecular characteristics of each cell type and can provide insights into their functions.</p>
<p>To find marker genes, we can use the <code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers()</a></code> function from the <code>scran</code> package. This function performs pairwise differential expression analysis between clusters or cell types. We can specify the clusters we want to compare and the method for calculating differential expression.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>    groups<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">level1class</span><span class="op">)</span><span class="op">)</span>, pval.type<span class="op">=</span><span class="st">"all"</span>,</span>
<span>    direction<span class="op">=</span><span class="st">"up"</span>, subset.row<span class="op">=</span><span class="va">top_var_genes</span>, test<span class="op">=</span><span class="st">"t"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers()</a></code> function identifies genes that are significantly upregulated in each cluster compared to all other clusters. The <code>pval.type="all"</code> argument specifies that we want to calculate p-values for all comparisons, and the <code>direction="up"</code> argument indicates that we are interested in genes that are upregulated in the specified clusters.</p>
<p>The <code>subset.row=top_var_genes</code> argument restricts the analysis to the highly variable genes we identified earlier, which are more likely to be informative for distinguishing cell types.</p>
<p>The resulting <code>markers</code> object is a list where each element corresponds to a cluster and contains the differential expression results for that cluster. We can visualize the top marker genes for each cluster using a heatmap.</p>
<p>Take my word for it that we can visualize the top marker genes for each cluster using a heatmap. This allows us to see which genes are most strongly associated with each cluster and how they differ across clusters.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># visualize the top marker gene for cluster 1 in a violin plot(</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="va">top_marker_genes_for_all_clusters</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">top_marker_genes_for_all_clusters</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    cluster_rows<span class="op">=</span><span class="cn">TRUE</span>, cluster_cols<span class="op">=</span><span class="cn">TRUE</span>, show_colnames<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    main<span class="op">=</span><span class="st">"Top marker genes for cluster 1"</span>,</span>
<span>    color_columns_by<span class="op">=</span><span class="st">"level1class"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="setup_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="interactive-visualization-with-isee" class="level2" data-number="28.6"><h2 data-number="28.6" class="anchored" data-anchor-id="interactive-visualization-with-isee">
<span class="header-section-number">28.6</span> Interactive Visualization with iSEE</h2>
<p>The <a href="https://www.bioconductor.org/packages/release/bioc/html/iSEE.html">Interactive SummarizedExperiment Explorer (iSEE)</a> is a powerful tool for exploring single-cell data interactively. It allows us to visualize and interact with the data in real-time, making it easier to explore complex datasets and identify patterns.</p>
<p>A key component of iSEE is the <code>SingleCellExperiment</code> object, which serves as the data container for single-cell RNA-seq data. iSEE provides a user-friendly interface for exploring the data.</p>
<p>To use iSEE, we first need to install the <code>iSEE</code> package and then launch the iSEE app.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Install iSEE package if not already installed</span></span>
<span><span class="fu">iSEE</span><span class="fu">::</span><span class="fu"><a href="https://isee.github.io/iSEE/reference/iSEE.html">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>iSEE()</code> function launches the iSEE app, which provides an interactive interface for exploring the <code>SingleCellExperiment</code> object.</p>
<p>You may want to play with a few genes to see how they are expressed across the different clusters. You can also explore the metadata associated with the cells, such as tissue type, cell type annotations, and quality control metrics. The set of “marker genes” we identified earlier can also be visualized in iSEE, allowing us to see how these genes are expressed across different clusters and cell types.</p>
<p>For cluster 1, the marker genes are:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">top_marker_genes_for_all_clusters</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="st">'\n'</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Aqp4 
 Cldn10 
 Clu 
 Aldoc 
 Ntsr2 
 Mt2 
 Fabp7 
 Gja1 
 Gpr37l1 
 Prdx6 </code></pre>
</div>
</div>
<p>See the</p>
</section><section id="bonus-working-with-10x-genomics-data" class="level2" data-number="28.7"><h2 data-number="28.7" class="anchored" data-anchor-id="bonus-working-with-10x-genomics-data">
<span class="header-section-number">28.7</span> Bonus: Working with 10x Genomics Data</h2>
<p>In the previous sections, we used a curated dataset from the <code>scRNAseq</code> package. However, many real-world single-cell RNA-seq datasets come from 10x Genomics platforms, which produce data in specific formats. In this section, we’ll cover a hypothetical workflow for loading and processing raw 10x Genomics data using R.</p>
<section id="understanding-10x-genomics-output-formats" class="level3" data-number="28.7.1"><h3 data-number="28.7.1" class="anchored" data-anchor-id="understanding-10x-genomics-output-formats">
<span class="header-section-number">28.7.1</span> Understanding 10x Genomics Output Formats</h3>
<p>10x Genomics Cell Ranger produces several output formats:</p>
<ol type="1">
<li>
<strong>Matrix Market format</strong> (<code>.mtx</code> files): Sparse matrix format with separate files for the matrix, gene names, and cell barcodes</li>
<li>
<strong>HDF5 format</strong> (<code>.h5</code> files): A single compressed file containing all the data</li>
<li>
<strong>Filtered vs.&nbsp;unfiltered</strong>: Cell Ranger provides both versions - filtered contains only cells that pass basic quality filters</li>
</ol></section><section id="loading-10x-data-from-matrix-market-files" class="level3" data-number="28.7.2"><h3 data-number="28.7.2" class="anchored" data-anchor-id="loading-10x-data-from-matrix-market-files">
<span class="header-section-number">28.7.2</span> Loading 10x Data from Matrix Market Files</h3>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Loading from standard 10x output directory</span></span>
<span><span class="co"># This assumes you have a directory with:</span></span>
<span><span class="co"># - matrix.mtx.gz (or matrix.mtx)</span></span>
<span><span class="co"># - features.tsv.gz (or genes.tsv.gz for older versions)</span></span>
<span><span class="co"># - barcodes.tsv.gz</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 1: Using DropletUtils (recommended)</span></span>
<span><span class="va">sce_10x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span><span class="op">(</span><span class="st">"path/to/10x/output/"</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 2: Using Seurat (if you prefer the Seurat ecosystem)</span></span>
<span><span class="co"># library(Seurat)</span></span>
<span><span class="co"># seurat_obj &lt;- Read10X("path/to/10x/output/")</span></span>
<span><span class="co"># sce_10x &lt;- as.SingleCellExperiment(CreateSeuratObject(seurat_obj))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>What we’re doing:</strong> The <code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts()</a></code> function from DropletUtils reads the three files that comprise 10x output and creates a SingleCellExperiment object. The <code>col.names = TRUE</code> parameter ensures that cell barcodes are used as column names.</p>
<p><strong>Why DropletUtils:</strong> This package is specifically designed for handling droplet-based single-cell data (like 10x Genomics). It includes specialized functions for:</p>
<ul>
<li>Reading 10x formats efficiently</li>
<li>Distinguishing empty droplets from cells</li>
<li>Handling ambient RNA contamination</li>
<li>Quality control specific to droplet-based methods</li>
</ul></section><section id="loading-from-hdf5-files" class="level3" data-number="28.7.3"><h3 data-number="28.7.3" class="anchored" data-anchor-id="loading-from-hdf5-files">
<span class="header-section-number">28.7.3</span> Loading from HDF5 Files</h3>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># For HDF5 format files</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/rhdf5">rhdf5</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 1: Using DropletUtils</span></span>
<span><span class="va">sce_h5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span><span class="op">(</span><span class="st">"path/to/sample.h5"</span>, type <span class="op">=</span> <span class="st">"HDF5"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 2: Direct HDF5 reading (more control)</span></span>
<span><span class="va">h5_file</span> <span class="op">&lt;-</span> <span class="st">"path/to/sample.h5"</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5_read.html">h5read</a></span><span class="op">(</span><span class="va">h5_file</span>, <span class="st">"matrix/features"</span><span class="op">)</span></span>
<span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5_read.html">h5read</a></span><span class="op">(</span><span class="va">h5_file</span>, <span class="st">"matrix/barcodes"</span><span class="op">)</span></span>
<span><span class="va">matrix_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5_read.html">h5read</a></span><span class="op">(</span><span class="va">h5_file</span>, <span class="st">"matrix/data"</span><span class="op">)</span></span>
<span><span class="co"># Additional processing needed to reconstruct sparse matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>What we’re doing:</strong> HDF5 files are more convenient as they contain all information in a single file. The <code>type = "HDF5"</code> parameter tells the function to expect this format instead of separate matrix files.</p>
<p><strong>When to use H5 files:</strong></p>
<ul>
<li>
<strong>Convenience</strong>: Single file is easier to manage and transfer</li>
<li>
<strong>Speed</strong>: Often faster to read than multiple compressed files</li>
<li>
<strong>Storage</strong>: More efficient storage, especially for large datasets</li>
</ul></section><section id="essential-quality-control-after-loading-raw-data" class="level3" data-number="28.7.4"><h3 data-number="28.7.4" class="anchored" data-anchor-id="essential-quality-control-after-loading-raw-data">
<span class="header-section-number">28.7.4</span> Essential Quality Control After Loading Raw Data</h3>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add mitochondrial gene information</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sce_10x</span><span class="op">)</span><span class="op">)</span> <span class="co"># Human mitochondrial genes</span></span>
<span><span class="co"># For mouse data, use: is.mito &lt;- grepl("^mt-", rownames(sce_10x))</span></span>
<span></span>
<span><span class="co"># Calculate per-cell QC metrics</span></span>
<span><span class="va">sce_10x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html">addPerCellQC</a></span><span class="op">(</span><span class="va">sce_10x</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Mito <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate per-gene QC metrics</span></span>
<span><span class="va">sce_10x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html">addPerFeatureQC</a></span><span class="op">(</span><span class="va">sce_10x</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the QC metrics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">sce_10x</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>What we’re doing:</strong> We’re calculating quality control metrics that are essential for raw 10x data:</p>
<ul>
<li>
<strong>Mitochondrial genes</strong>: High mitochondrial gene expression often indicates dying cells</li>
<li>
<strong>Per-cell metrics</strong>: Total counts, number of detected genes, mitochondrial percentage</li>
<li>
<strong>Per-gene metrics</strong>: How many cells express each gene</li>
</ul>
<p><strong>Why QC is crucial for raw data:</strong> Unlike curated datasets, raw 10x data contains:</p>
<ul>
<li>
<strong>Empty droplets</strong>: Droplets that captured ambient RNA but no cells</li>
<li>
<strong>Dying cells</strong>: Cells with compromised membranes</li>
<li>
<strong>Doublets</strong>: Droplets containing multiple cells</li>
<li>
<strong>Low-quality cells</strong>: Cells with very few detected genes</li>
</ul></section><section id="filtering-low-quality-cells-and-genes" class="level3" data-number="28.7.5"><h3 data-number="28.7.5" class="anchored" data-anchor-id="filtering-low-quality-cells-and-genes">
<span class="header-section-number">28.7.5</span> Filtering Low-Quality Cells and Genes</h3>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize QC metrics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span><span class="op">(</span><span class="va">sce_10x</span>, x <span class="op">=</span> <span class="st">"sum"</span>, y <span class="op">=</span> <span class="st">"detected"</span>, colour_by <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set filtering thresholds (these are examples - adjust based on your data)</span></span>
<span><span class="co"># Filter cells</span></span>
<span><span class="va">high_mito</span> <span class="op">&lt;-</span> <span class="va">sce_10x</span><span class="op">$</span><span class="va">subsets_Mito_percent</span> <span class="op">&gt;</span> <span class="fl">20</span>  <span class="co"># Remove cells with &gt;20% mitochondrial reads</span></span>
<span><span class="va">low_lib_size</span> <span class="op">&lt;-</span> <span class="va">sce_10x</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;</span> <span class="fl">1000</span>              <span class="co"># Remove cells with &lt;1000 total counts</span></span>
<span><span class="va">low_n_features</span> <span class="op">&lt;-</span> <span class="va">sce_10x</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;</span> <span class="fl">500</span>        <span class="co"># Remove cells with &lt;500 detected genes</span></span>
<span></span>
<span><span class="co"># Filter genes (remove genes expressed in very few cells)</span></span>
<span><span class="va">low_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html">counts</a></span><span class="op">(</span><span class="va">sce_10x</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">10</span>  <span class="co"># Expressed in &lt;10 cells</span></span>
<span></span>
<span><span class="co"># Apply filters</span></span>
<span><span class="va">sce_filtered</span> <span class="op">&lt;-</span> <span class="va">sce_10x</span><span class="op">[</span><span class="op">!</span><span class="va">low_expression</span>, <span class="op">!</span><span class="op">(</span><span class="va">high_mito</span> <span class="op">|</span> <span class="va">low_lib_size</span> <span class="op">|</span> <span class="va">low_n_features</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Check how many cells and genes remain</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sce_10x</span><span class="op">)</span>      <span class="co"># Before filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sce_filtered</span><span class="op">)</span> <span class="co"># After filtering</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>What we’re doing:</strong> We’re applying filters to remove low-quality cells and rarely expressed genes. The specific thresholds should be adjusted based on your dataset characteristics.</p>
<p><strong>Understanding the filters:</strong></p>
<ul>
<li>
<strong>Mitochondrial percentage</strong>: Cells with high mitochondrial gene expression are likely dying</li>
<li>
<strong>Library size</strong>: Total number of UMIs per cell - very low values suggest poor capture</li>
<li>
<strong>Number of features</strong>: How many different genes are detected - very low values suggest poor quality</li>
<li>
<strong>Gene expression frequency</strong>: Genes expressed in very few cells are often noise</li>
</ul>
<p><strong>Critical considerations:</strong></p>
<ul>
<li>
<strong>Thresholds are dataset-dependent</strong>: What works for one experiment may not work for another</li>
<li>
<strong>Biological vs.&nbsp;technical variation</strong>: Some cell types naturally have different RNA content</li>
<li>
<strong>Visualization first</strong>: Always plot your QC metrics before setting thresholds</li>
</ul></section></section><section id="further-reading-and-resources" class="level2" data-number="28.8"><h2 data-number="28.8" class="anchored" data-anchor-id="further-reading-and-resources">
<span class="header-section-number">28.8</span> Further Reading and Resources</h2>
<p>For more in-depth information on single-cell RNA-seq analysis, consider the following resources:</p>
<ul>
<li>The <a href="https://osca.bioconductor.org/">Orchestrating Single-Cell RNA-seq Analysis</a> book provides a comprehensive guide to single-cell RNA-seq analysis using Bioconductor.</li>
<li>The <a href="https://satijalab.org/seurat/">Seurat</a> package documentation offers extensive tutorials and vignettes for single-cell analysis in the Seurat ecosystem.</li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-lun2016" class="csl-entry" role="listitem">
Lun, Aaron T. L., Davis J. McCarthy, and John C. Marioni. 2016. <span>“A Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data with Bioconductor.”</span> <em>F1000Res.</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-McCarthy2017scater" class="csl-entry" role="listitem">
McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Willis. 2017. <span>“Scater: Pre-Processing, Quality Control, Normalisation and Visualisation of Single-Cell <span>R</span><span>N</span><span>A</span>-Seq Data in <span>R</span>.”</span> <em>Bioinformatics</em> 33: 1179–86. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.
</div>
<div id="ref-risso2024scRNAseq" class="csl-entry" role="listitem">
Risso, Davide, and Michael Cole. 2024. <em>scRNAseq: Collection of Public Single-Cell RNA-Seq Datasets</em>. <a href="https://doi.org/10.18129/B9.bioc.scRNAseq">https://doi.org/10.18129/B9.bioc.scRNAseq</a>.
</div>
<div id="ref-zeiselCellTypesMouse2015" class="csl-entry" role="listitem">
Zeisel, Amit, Ana B. Muñoz-Manchado, Simone Codeluppi, Peter Lönnerberg, Gioele La Manno, Anna Juréus, Sueli Marques, et al. 2015. <span>“Cell Types in the Mouse Cortex and Hippocampus Revealed by Single-Cell <span class="nocase">RNA-seq</span>.”</span> <em>Science</em> 347 (6226): 1138–42. <a href="https://doi.org/10.1126/science.aaa1934">https://doi.org/10.1126/science.aaa1934</a>.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../atac-seq/atac-seq.html" class="pagination-link" aria-label="Genomic ranges and ATAC-Seq">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Genomic ranges and ATAC-Seq</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
    <a class="nav-link" href="../license.html">
<p>License</p>
</a>
  </li>  
</ul>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/seandavi/RBiocBook/edit/main/single_cell/setup.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/seandavi/RBiocBook/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/seandavi/RBiocBook/blob/main/single_cell/setup.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>